In [1]:
import warnings
warnings.filterwarnings('ignore')

## (Done)과제

### Q1. titanic 데이터셋에서 age, fare 2개 열을 선택하여 데이터프레임 df를 만드세요

In [2]:
import pandas as pd
import seaborn as sns

df = sns.load_dataset('titanic')

In [3]:
df = pd.concat([df.age, df.fare], axis= 1)
df.head()

age     fare
0  22.0   7.2500
1  38.0  71.2833
2  26.0   7.9250
3  35.0  53.1000
4  35.0   8.0500

### Q2. 10을 더하는 함수 add_10 과 두 객체의 합 add_two_obj 함수를 정의한 후 결과값이 각각 20과 20을 출력하세요.


In [4]:
def add_10(x):
    return x + 10

def add_two_obj(a, b):
    return a + b

print(add_10(10))
print(add_two_obj(15, 5))

20
20


### Q3. 10을 곱하는 사용자 함수와 두 객체의 나눗셈을 수행하는 사용자 함수를 정의한 후 결과값이 각각 200과 10을 출력하세요.


In [5]:
def mul_10(x):
    return x*10

def div(a, b):
    try:
        return a/b
    except ZeroDivisionError as ze:
        print(ze)
print(mul_10(20))
print(div(100, 10))
        
        

200
10.0


### Q4. df.age 에 add_10 함수를 적용해서 출력하세요.


In [6]:
df.head()

age     fare
0  22.0   7.2500
1  38.0  71.2833
2  26.0   7.9250
3  35.0  53.1000
4  35.0   8.0500

In [7]:
print(df['age'].apply(add_10).head())

0    32.0
1    48.0
2    36.0
3    45.0
4    45.0
Name: age, dtype: float64


### Q5. 시리즈 객체 + 숫자를 연산하는 방식으로 df.age에 10을 더한 후 sr2에 저장하고 출력하세요.


In [8]:
sr2 = df.age + 10
print(sr2.head())

0    32.0
1    48.0
2    36.0
3    45.0
4    45.0
Name: age, dtype: float64


### Q6. 람다 함수를 활용해서 위와 같은 연산을 수행하고 결과를 sr3로 출력하세요.


In [9]:
sr3 = df['age'].apply(lambda x : x + 10)
print(sr3.head())

0    32.0
1    48.0
2    36.0
3    45.0
4    45.0
Name: age, dtype: float64


### Q7. df의 'age', 'fare' 칼럼에 10을 덧셈연산하는 사용자 함수를 정의하고 applymap()으로 각 사용자 함수를 매핑한 후 출력하세요. (결과 dataframe 이름이 df_map)


In [10]:
df_map = df.applymap(add_10)
df_map.head()

age     fare
0  32.0  17.2500
1  48.0  81.2833
2  36.0  17.9250
3  45.0  63.1000
4  45.0  18.0500

### Q8. df_map의 [0,0]을 NaN으로 변경한 후 missing_value 함수를 적용해서 불린 시리즈를 반환하세요.


In [11]:
import numpy as np
df_map.iloc[0,0] = np.nan
df_map.head()

age     fare
0   NaN  17.2500
1  48.0  81.2833
2  36.0  17.9250
3  45.0  63.1000
4  45.0  18.0500

In [12]:
type(df_map.iloc[0,0])

numpy.float64

In [13]:
df_map.isna().head()

age   fare
0   True  False
1  False  False
2  False  False
3  False  False
4  False  False

In [14]:
def missing_value(sr):
    return sr.isnull()
print(missing_value(df_map['age']).head())
print(missing_value(df_map['fare']).head())

0     True
1    False
2    False
3    False
4    False
Name: age, dtype: bool
0    False
1    False
2    False
3    False
4    False
Name: fare, dtype: bool


### Q9. titanic 데이터셋을 seaborn에서 로딩한 후 블린 인덱싱을 이용하여 아래 사항들을 수행하세요.
- 나이가 10대(10~19세)인 승객만 따로 선택
- 나이가 10세 미만(0~9세)이고 여성인 승객만 따로 선택
- 나이가 10세 미만(0~9세) 또는 60세 이상인 승객의 age, sex, alone 열만 선택


In [15]:
import pandas as pd
import seaborn as sns

df_ttn = sns.load_dataset('titanic')

In [16]:
df_ttn[(df_ttn.age >=10) & (df_ttn.age <=19)].head() # and /or 사용 안됨. & /|

survived  pclass     sex   age  sibsp  parch      fare embarked   class  \
9          1       2  female  14.0      1      0   30.0708        C  Second   
14         0       3  female  14.0      0      0    7.8542        S   Third   
22         1       3  female  15.0      0      0    8.0292        Q   Third   
27         0       1    male  19.0      3      2  263.0000        S   First   
38         0       3  female  18.0      2      0   18.0000        S   Third   

      who  adult_male deck  embark_town alive  alone  
9   child       False  NaN    Cherbourg   yes  False  
14  child       False  NaN  Southampton    no   True  
22  child       False  NaN   Queenstown   yes   True  
27    man        True    C  Southampton    no  False  
38  woman       False  NaN  Southampton    no  False

In [17]:
((df_ttn.age >=10) & (df_ttn.age <=19)).head() # 이 조건문 자체가 boolean 시리즈

0    False
1    False
2    False
3    False
4    False
Name: age, dtype: bool

In [18]:
df_ttn[(df_ttn.age <= 9) & (df_ttn.sex == 'female')].head()

survived  pclass     sex  age  sibsp  parch     fare embarked   class  \
10          1       3  female  4.0      1      1  16.7000        S   Third   
24          0       3  female  8.0      3      1  21.0750        S   Third   
43          1       2  female  3.0      1      2  41.5792        C  Second   
58          1       2  female  5.0      1      2  27.7500        S  Second   
119         0       3  female  2.0      4      2  31.2750        S   Third   

       who  adult_male deck  embark_town alive  alone  
10   child       False    G  Southampton   yes  False  
24   child       False  NaN  Southampton    no  False  
43   child       False  NaN    Cherbourg   yes  False  
58   child       False  NaN  Southampton   yes  False  
119  child       False  NaN  Southampton    no  False

In [19]:
df_ttn[(df_ttn.age <= 9) | (df_ttn.age >=60)].loc[:, ['age', 'sex', 'alone']].head()

age     sex  alone
7    2.0    male  False
10   4.0  female  False
16   2.0    male  False
24   8.0  female  False
33  66.0    male   True

In [20]:
df_ttn[(df_ttn.age <= 9) | (df_ttn.age >=60)][['age', 'sex', 'alone']].head()  # .loc을 안쓴다면 []안에 list_column_names 만 넣어도 됨

age     sex  alone
7    2.0    male  False
10   4.0  female  False
16   2.0    male  False
24   8.0  female  False
33  66.0    male   True

### Q10. 나이가 10세 미만(0~9세) 또는 60세 이상인 승객의 alone 열만 선택한 후 값을 True로 변경하세요


이게 원래 df_ttn에 반영되지 않으면 의미가 없을 것 같다?

In [21]:
idx = df_ttn[(df_ttn.age <= 9) | (df_ttn.age >=60)].index
idx

Int64Index([  7,  10,  16,  24,  33,  43,  50,  54,  58,  63,  78,  96, 116,
            119, 147, 164, 165, 170, 171, 172, 182, 183, 184, 193, 205, 233,
            237, 252, 261, 275, 278, 280, 297, 305, 326, 340, 348, 366, 374,
            381, 386, 407, 438, 445, 448, 456, 469, 479, 480, 483, 489, 493,
            530, 535, 541, 545, 549, 555, 570, 587, 618, 625, 630, 634, 642,
            644, 672, 684, 691, 694, 720, 745, 750, 751, 755, 777, 787, 788,
            803, 813, 824, 827, 829, 831, 850, 851, 852, 869],
           dtype='int64')

In [22]:
df_ttn.loc[idx, 'alone'] = True

In [23]:
set(df_ttn.loc[idx, 'alone'])

{True}

## (Done) 과제 #1 Fri 0701

### Q11. 최대값 - 최소값을 반환하는 사용자 함수를 min_max로 정의하고 age, fare 열로 구성되는 타이타닉 데이터프레임df 에 적용하여 결과와 자료형을 출력하세요. 
source : titanic = sns.load_dataset('titanic')


In [24]:
titanic = sns.load_dataset('titanic')

In [25]:
df = pd.concat([titanic.age, titanic.fare], axis = 1)
df.head()

age     fare
0  22.0   7.2500
1  38.0  71.2833
2  26.0   7.9250
3  35.0  53.1000
4  35.0   8.0500

In [26]:
def min_max(x):
    return max(x) - min(x)

In [27]:
result = df.apply(min_max)
print(result, type(result))

age      79.5800
fare    512.3292
dtype: float64 <class 'pandas.core.series.Series'>


### Q12. df 에서 'age' 칼럼과 'fare'칼럼을 더한 가격을 리턴하는 사용자 함수를 정의한 후 df['add'] 변수를 생성하여 출력하세요.(lambda 함수 사용)


In [28]:
def add(x):
    return x.age + x.fare
df['add'] = add(df)
df.head()

age     fare       add
0  22.0   7.2500   29.2500
1  38.0  71.2833  109.2833
2  26.0   7.9250   33.9250
3  35.0  53.1000   88.1000
4  35.0   8.0500   43.0500

In [29]:
df['add_1'] = df.age + df.fare
df.head()

age     fare       add     add_1
0  22.0   7.2500   29.2500   29.2500
1  38.0  71.2833  109.2833  109.2833
2  26.0   7.9250   33.9250   33.9250
3  35.0  53.1000   88.1000   88.1000
4  35.0   8.0500   43.0500   43.0500

### Q13. 'age', 'fare' 두개 칼럼으로 구성되는 titanic 데이터프레임에서 [0,1]의 값을 NaN을 변경한 후 df 이름으로 출력하세요.


In [30]:
df = pd.concat([titanic.age, titanic.fare], axis = 1)
df.head()

age     fare
0  22.0   7.2500
1  38.0  71.2833
2  26.0   7.9250
3  35.0  53.1000
4  35.0   8.0500

In [31]:
df.iloc[0, 1] = np.nan
df.head()

age     fare
0  22.0      NaN
1  38.0  71.2833
2  26.0   7.9250
3  35.0  53.1000
4  35.0   8.0500

### Q14. 각 열의 NaN 찾기, 각 열의 NaN 개수 반환, 데이터프레임의 총 NaN 개수를 반환하는 사용자 함수를 정의한 후 df.pipe 메소드로 함수 매핑한 결과를 모두 출력하세요.


In [32]:
def nan_info(x):
    list_nan_idx = []
    list_nan_count = []
    
    for col in x.columns:
        x_na = x[col].isna()
        idx = x_na[x_na].index
        list_nan_idx.append(idx)
        
        list_nan_count.append(x[col].isna().sum())
    
    nan_all = x.isnull().sum().sum()
        
    return list_nan_idx, list_nan_count, nan_all
df.pipe(nan_info)

([Int64Index([  5,  17,  19,  26,  28,  29,  31,  32,  36,  42,
              ...
              832, 837, 839, 846, 849, 859, 863, 868, 878, 888],
             dtype='int64', length=177),
  Int64Index([0], dtype='int64')],
 [177, 1],
 178)

### Q15. 타이타닉 df의 열 순서를 아래와 같이 재구성하고 출력하세요
* 열 선택 : ['pclass', 'sex', 'age', 'survived'].
* 열 이름의 리스트 만들기
* 열 이름을 알파벳 순으로 정렬하기
* 열 이름을 기존 순서의 정반대 역순으로 정렬하기
* 열 이름을 사용자가 정의한 임의의 순서로 재배치하기


In [33]:
df = titanic[['pclass', 'sex', 'age', 'survived']]
df.head()

pclass     sex   age  survived
0       3    male  22.0         0
1       1  female  38.0         1
2       3  female  26.0         1
3       1  female  35.0         1
4       3    male  35.0         0

In [34]:
list_col = list(df.columns)
list_col

['pclass', 'sex', 'age', 'survived']

In [35]:
list_col_sorted = list_col.copy()
list_col_sorted.sort()
list_col_sorted

['age', 'pclass', 'sex', 'survived']

In [36]:
df = df[list_col_sorted]
df.head()

age  pclass     sex  survived
0  22.0       3    male         0
1  38.0       1  female         1
2  26.0       3  female         1
3  35.0       1  female         1
4  35.0       3    male         0

In [37]:
list_col_rev = list_col.copy()
list_col_rev.reverse()
df = df[list_col_rev]
df.head()

survived   age     sex  pclass
0         0  22.0    male       3
1         1  38.0  female       1
2         1  26.0  female       3
3         1  35.0  female       1
4         0  35.0    male       3

In [38]:
df = df[['pclass', 'age', 'sex', 'survived']]
df.head()

pclass   age     sex  survived
0       3  22.0    male         0
1       1  38.0  female         1
2       3  26.0  female         1
3       1  35.0  female         1
4       3  35.0    male         0

# 그룹 연산

<img src='dataset/그룹연산.png' STYLE='width:800px;'>

In [39]:
import pandas as pd
import seaborn as sns

titanic = sns.load_dataset('titanic')
titanic.columns

Index(['survived', 'pclass', 'sex', 'age', 'sibsp', 'parch', 'fare',
       'embarked', 'class', 'who', 'adult_male', 'deck', 'embark_town',
       'alive', 'alone'],
      dtype='object')

In [40]:
df = titanic.loc[:, ['age', 'sex', 'class', 'fare', 'survived']]

print('승객 수:', len(df))
print(df.head())

승객 수: 891
    age     sex  class     fare  survived
0  22.0    male  Third   7.2500         0
1  38.0  female  First  71.2833         1
2  26.0  female  Third   7.9250         1
3  35.0  female  First  53.1000         1
4  35.0    male  Third   8.0500         0


## 그룹핑

In [41]:
df.columns, type(df.columns)

(Index(['age', 'sex', 'class', 'fare', 'survived'], dtype='object'),
 pandas.core.indexes.base.Index)

In [42]:
grouped = df.groupby(['class'])

In [43]:
type(grouped)# GroupBy object, list로 바꿔서 접근하는 방법이 있음

pandas.core.groupby.generic.DataFrameGroupBy

In [44]:
list_g = list(grouped)
type(list_g), type(list_g[0]), type(list_g[0][0]), type(list_g[0][1])

(list, tuple, str, pandas.core.frame.DataFrame)

In [45]:
print(grouped)
list(grouped)

[('First',
        age     sex  class     fare  survived
  1    38.0  female  First  71.2833         1
  3    35.0  female  First  53.1000         1
  6    54.0    male  First  51.8625         0
  11   58.0  female  First  26.5500         1
  23   28.0    male  First  35.5000         1
  ..    ...     ...    ...      ...       ...
  871  47.0  female  First  52.5542         1
  872  33.0    male  First   5.0000         0
  879  56.0  female  First  83.1583         1
  887  19.0  female  First  30.0000         1
  889  26.0    male  First  30.0000         1
  
  [216 rows x 5 columns]),
 ('Second',
        age     sex   class     fare  survived
  9    14.0  female  Second  30.0708         1
  15   55.0  female  Second  16.0000         1
  17    NaN    male  Second  13.0000         1
  20   35.0    male  Second  26.0000         0
  21   34.0    male  Second  13.0000         1
  ..    ...     ...     ...      ...       ...
  866  27.0  female  Second  13.8583         1
  874  28.0  female

## 그룹객체를 iteration 으로 출력

In [46]:
for key, group in grouped:
    print('* key : ', key)
    print('*number : ', len(group))
    print(group.head(), '\n')

* key :  First
*number :  216
     age     sex  class     fare  survived
1   38.0  female  First  71.2833         1
3   35.0  female  First  53.1000         1
6   54.0    male  First  51.8625         0
11  58.0  female  First  26.5500         1
23  28.0    male  First  35.5000         1 

* key :  Second
*number :  184
     age     sex   class     fare  survived
9   14.0  female  Second  30.0708         1
15  55.0  female  Second  16.0000         1
17   NaN    male  Second  13.0000         1
20  35.0    male  Second  26.0000         0
21  34.0    male  Second  13.0000         1 

* key :  Third
*number :  491
    age     sex  class     fare  survived
0  22.0    male  Third   7.2500         0
2  26.0  female  Third   7.9250         1
4  35.0    male  Third   8.0500         0
5   NaN    male  Third   8.4583         0
7   2.0    male  Third  21.0750         0 



## 연산 메서드 적용

In [47]:
average = grouped.mean()
print(type(average))
average

<class 'pandas.core.frame.DataFrame'>


age       fare  survived
class                                 
First   38.233441  84.154687  0.629630
Second  29.877630  20.662183  0.472826
Third   25.140620  13.675550  0.242363

In [48]:
stat = grouped.max()
stat

age   sex      fare  survived
class                                 
First   80.0  male  512.3292         1
Second  70.0  male   73.5000         1
Third   74.0  male   69.5500         1

In [49]:
# Groupby_object. 하고 min, median, mean, max 있는 것 확인

In [50]:
# Q. 3rd 그룹만을 선택새거 group3 이름으로 저장하고 통계 요약표를 출력하세요.

In [51]:
group3 = grouped.get_group('Third')
group3.describe()

age        fare    survived
count  355.000000  491.000000  491.000000
mean    25.140620   13.675550    0.242363
std     12.495398   11.778142    0.428949
min      0.420000    0.000000    0.000000
25%     18.000000    7.750000    0.000000
50%     24.000000    8.050000    0.000000
75%     32.000000   15.500000    0.000000
max     74.000000   69.550000    1.000000

In [52]:
grouped_two = df.groupby(['class', 'sex'])
average_two = grouped_two.mean()
average_two

age        fare  survived
class  sex                                    
First  female  34.611765  106.125798  0.968085
       male    41.281386   67.226127  0.368852
Second female  28.722973   21.970121  0.921053
       male    30.740707   19.741782  0.157407
Third  female  21.750000   16.118810  0.500000
       male    26.507589   12.661633  0.135447

In [53]:
group3f = grouped_two.get_group(('Third', 'female'))
group3f.head()

age     sex  class     fare  survived
2   26.0  female  Third   7.9250         1
8   27.0  female  Third  11.1333         1
10   4.0  female  Third  16.7000         1
14  14.0  female  Third   7.8542         0
18  31.0  female  Third  18.0000         0

## 모든 열에 여러 함수를 매핑

In [54]:
agg_all = grouped.agg(['min', 'max']) #list_function 을 넣는다.
agg_all.head()

age           sex       fare           survived    
         min   max     min   max  min       max      min max
class                                                       
First   0.92  80.0  female  male  0.0  512.3292        0   1
Second  0.67  70.0  female  male  0.0   73.5000        0   1
Third   0.42  74.0  female  male  0.0   69.5500        0   1

In [55]:
agg_all = grouped.agg(['min', 'max', 'mean', 'median']) #list_function 을 넣는다.
agg_all

age                         fare                                \
         min   max       mean median  min       max       mean   median   
class                                                                     
First   0.92  80.0  38.233441   37.0  0.0  512.3292  84.154687  60.2875   
Second  0.67  70.0  29.877630   29.0  0.0   73.5000  20.662183  14.2500   
Third   0.42  74.0  25.140620   24.0  0.0   69.5500  13.675550   8.0500   

       survived                       
            min max      mean median  
class                                 
First         0   1  0.629630    1.0  
Second        0   1  0.472826    0.0  
Third         0   1  0.242363    0.0

## 각 열마다 다른 함수를 적용하여 집계

In [56]:
agg_sep = grouped.agg({'fare': ['min', 'max'], 'age': 'mean'}) 
# dictionaty 형태로 넣는다. column_name: list_function|function
agg_sep.head()

fare                  age
        min       max       mean
class                           
First   0.0  512.3292  38.233441
Second  0.0   73.5000  29.877630
Third   0.0   69.5500  25.140620

## 그룹을 filtering

In [57]:
len(grouped.get_group('First')), grouped.get_group('First')

(216,
       age     sex  class     fare  survived
 1    38.0  female  First  71.2833         1
 3    35.0  female  First  53.1000         1
 6    54.0    male  First  51.8625         0
 11   58.0  female  First  26.5500         1
 23   28.0    male  First  35.5000         1
 ..    ...     ...    ...      ...       ...
 871  47.0  female  First  52.5542         1
 872  33.0    male  First   5.0000         0
 879  56.0  female  First  83.1583         1
 887  19.0  female  First  30.0000         1
 889  26.0    male  First  30.0000         1
 
 [216 rows x 5 columns])

In [58]:
grouped_filter = grouped.filter(lambda x : len(x) >= 200)
grouped_filter.head()

age     sex  class     fare  survived
0  22.0    male  Third   7.2500         0
1  38.0  female  First  71.2833         1
2  26.0  female  Third   7.9250         1
3  35.0  female  First  53.1000         1
4  35.0    male  Third   8.0500         0

In [59]:
grouped_filter.head() # filter 된 행들을 다 묶어놓음, DataFrame

age     sex  class     fare  survived
0  22.0    male  Third   7.2500         0
1  38.0  female  First  71.2833         1
2  26.0  female  Third   7.9250         1
3  35.0  female  First  53.1000         1
4  35.0    male  Third   8.0500         0

In [60]:
type(grouped_filter)

pandas.core.frame.DataFrame

In [61]:
grouped_filter['class'].value_counts() #category object 이라서 없는 Second도 보여줌

Third     491
First     216
Second      0
Name: class, dtype: int64

In [62]:
grouped_filter['class'].values, set(grouped_filter['class'].values)

(['Third', 'First', 'Third', 'First', 'Third', ..., 'Third', 'First', 'Third', 'First', 'Third']
 Length: 707
 Categories (3, object): ['First', 'Second', 'Third'],
 {'First', 'Third'})

Q 데이터 개수가 200개 이상인 그룹만을 필터링하여 데이터프레임으로 반환하여 grouped_filter 
이름으로 저장 후 처음 5개행을 출력하세요.

In [63]:
grouped_filter = grouped.filter(lambda x : len(x) >= 200)
grouped_filter.head()

age     sex  class     fare  survived
0  22.0    male  Third   7.2500         0
1  38.0  female  First  71.2833         1
2  26.0  female  Third   7.9250         1
3  35.0  female  First  53.1000         1
4  35.0    male  Third   8.0500         0

## (Done) 과제 #2 Fri 0701
Q age 열의 평균이 30보다 작은 그룹만을 필터링하여 데이터프레임으로 반환하여 age_filter 이름으로 저장후 처음 5개행 출력

In [64]:
tt = sns.load_dataset('titanic')
tt.head()

survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True

In [65]:
group_tt = tt.groupby(['class', 'sex'])
list(group_tt)

[(('First', 'female'),
       survived  pclass     sex   age  sibsp  parch      fare embarked  class  \
  1           1       1  female  38.0      1      0   71.2833        C  First   
  3           1       1  female  35.0      1      0   53.1000        S  First   
  11          1       1  female  58.0      0      0   26.5500        S  First   
  31          1       1  female   NaN      1      0  146.5208        C  First   
  52          1       1  female  49.0      1      0   76.7292        C  First   
  ..        ...     ...     ...   ...    ...    ...       ...      ...    ...   
  856         1       1  female  45.0      1      1  164.8667        S  First   
  862         1       1  female  48.0      0      0   25.9292        S  First   
  871         1       1  female  47.0      1      1   52.5542        S  First   
  879         1       1  female  56.0      0      1   83.1583        C  First   
  887         1       1  female  19.0      0      0   30.0000        S  First   
  
  

In [66]:
age_filter = group_tt.filter(lambda x : np.mean(x.age) < 30)

- 되었나 체크

In [67]:
for key, group in group_tt:
    mean_age = np.mean(group.age)
    print('* key : ', key, '\t*mean age : %.2f'%mean_age, end = '\t')
    if mean_age < 30:
        print('Under 30', end = '\n')
    else:
        print('\t')
# 3rd all, 2nd female 이 filter되어야 함

* key :  ('First', 'female') 	*mean age : 34.61		
* key :  ('First', 'male') 	*mean age : 41.28		
* key :  ('Second', 'female') 	*mean age : 28.72	Under 30
* key :  ('Second', 'male') 	*mean age : 30.74		
* key :  ('Third', 'female') 	*mean age : 21.75	Under 30
* key :  ('Third', 'male') 	*mean age : 26.51	Under 30


In [68]:
test_g = age_filter.groupby(['class', 'sex'])
for key, group in test_g:
    print(key)

('Second', 'female')
('Third', 'female')
('Third', 'male')


# pivot
```
피벗 테이블을 구성하는 4가지 요소
  - 행인덱스
  - 열인덱스
  - 데이터 값
  - 데이터 집계함수
 에 적용할 데이터프레임의 열을 각각 지정하여 함수의 인자로 전달
```

In [69]:
import seaborn as sns

titanic = sns.load_dataset('titanic')
df = titanic.loc[:, ['age', 'sex', 'class', 'fare', 'survived']]
df.head()

age     sex  class     fare  survived
0  22.0    male  Third   7.2500         0
1  38.0  female  First  71.2833         1
2  26.0  female  Third   7.9250         1
3  35.0  female  First  53.1000         1
4  35.0    male  Third   8.0500         0

In [70]:
pdf1 = pd.pivot_table(df,
                     index= 'class',
                     columns='sex',
                     values = 'age',
                     aggfunc = 'mean')
pdf1

# index X columns 

sex        female       male
class                       
First   34.611765  41.281386
Second  28.722973  30.740707
Third   21.750000  26.507589

In [71]:
pdf1 = pd.pivot_table(df,
                     index= 'sex',
                     columns='class',
                     values = 'age',
                     aggfunc = 'mean')
pdf1

# index X columns 

class       First     Second      Third
sex                                    
female  34.611765  28.722973  21.750000
male    41.281386  30.740707  26.507589

In [72]:
pdf1 = pd.pivot_table(df,
                     index= 'class',
                     columns='sex',
                     values = 'age',
                     aggfunc = 'min')
pdf1

# index X columns 

sex     female  male
class               
First     2.00  0.92
Second    2.00  0.67
Third     0.75  0.42

## 복수 인자 지정

In [73]:
pdf2 = pd.pivot_table(df,
                     index = ['class', 'sex'],
                     columns = 'survived',
                     values = ['age', 'fare'],
                     aggfunc = ['mean', 'max'])
pdf2

# 출력물은 역순으로

mean                                      max        \
                     age                   fare               age         
survived               0          1           0           1     0     1   
class  sex                                                                
First  female  25.666667  34.939024  110.604167  105.978159  50.0  63.0   
       male    44.581967  36.248000   62.894910   74.637320  71.0  80.0   
Second female  36.000000  28.080882   18.250000   22.288989  57.0  55.0   
       male    33.369048  16.022000   19.488965   21.095100  70.0  62.0   
Third  female  23.818182  19.329787   19.773093   12.464526  48.0  63.0   
       male    27.255814  22.274211   12.204469   15.579696  74.0  45.0   

                                 
                 fare            
survived            0         1  
class  sex                       
First  female  151.55  512.3292  
       male    263.00  512.3292  
Second female   26.00   65.0000  
       male     73.50   39.0000  
Third  female   69.55   31.3875  
       male     69.55   56.4958

# 결측치

In [74]:
import seaborn as sns

df_t = sns.load_dataset('titanic')
df_t.isnull().sum()

survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64

In [75]:
df_t.embarked.value_counts()

S    644
C    168
Q     77
Name: embarked, dtype: int64

## 앞의 값으로 채우기

In [76]:
df_t.embarked.fillna(method='ffill', inplace=True)
df_t.embarked.value_counts()

S    644
C    169
Q     78
Name: embarked, dtype: int64

In [77]:
df_t.isna().sum()

survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         0
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64

## (★Done)과제 #3 Fri 0701
 - age의 결측치를 적절한 값으로 대체하고 근거 설명
 - deck 컬럼의 삭제여부를 판단하는 근거 설명

```
df.dropna(subset=['column'], inplace=True)
df.dropna(inplace=True)
df.column.replace(np.nan, df.column.mean(), inplace=True)

참고
```

### age의 결측치 대체 & 근거

 - 그루핑 후 각 그룹의 중간값으로 대체할 것

#### dataset 정리

In [78]:
df_tt = sns.load_dataset('titanic')
df_tt.head()
# C = Cherbourg, Q = Queenstown, S = Southampton
# 프랑스, 영국?, 영국

survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True

In [79]:
df_tt.columns

Index(['survived', 'pclass', 'sex', 'age', 'sibsp', 'parch', 'fare',
       'embarked', 'class', 'who', 'adult_male', 'deck', 'embark_town',
       'alive', 'alone'],
      dtype='object')

In [80]:
# family 추가
df_tt['family'] = df_tt.sibsp + df_tt.parch
df_tt.head()

survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  family  
0    man        True  NaN  Southampton    no  False       1  
1  woman       False    C    Cherbourg   yes  False       1  
2  woman       False  NaN  Southampton   yes   True       0  
3  woman       False    C  Southampton   yes  False       1  
4    man        True  NaN  Southampton    no   True       0

In [81]:
set(df_tt['deck'])
# 선실

{'A', 'B', 'C', 'D', 'E', 'F', 'G', nan}

In [82]:
# 중복 항목 드롭
df_ttt = df_tt.copy()
df_ttt = df_ttt.drop(columns=['sibsp','parch', 'class', 'who', 'adult_male', 'embark_town', 'alive', 'alone'])
df_ttt.head()

survived  pclass     sex   age     fare embarked deck  family
0         0       3    male  22.0   7.2500        S  NaN       1
1         1       1  female  38.0  71.2833        C    C       1
2         1       3  female  26.0   7.9250        S  NaN       0
3         1       1  female  35.0  53.1000        S    C       1
4         0       3    male  35.0   8.0500        S  NaN       0

#### age 결측치를 채우기 위한 gouping에 사용할 column 결정

- 제외
    - survived
    - fare(pclass로 대변될 수 있다고 생각+ 사용하려면 구간 나눠야 하는데 유의미한 분류방법 모름), 
    - deck (1/2/3 등급이 각각 -E/ D-F/ D-G로, pclass와의 연관성이 크다 생각. (중복))
    - family
    - embarked

- 포함
    - pclass
    - sex

In [83]:
set(df_ttt['sex'])

{'female', 'male'}

In [84]:
set(df_ttt['pclass'])

{1, 2, 3}

In [85]:
g_test = df_ttt.groupby(['pclass', 'sex'])
g_med = g_test['age'].median()
print(g_med)

pclass  sex   
1       female    35.0
        male      40.0
2       female    28.0
        male      30.0
3       female    21.5
        male      25.0
Name: age, dtype: float64


In [86]:
nan_age = df_ttt['age'].isna()
nan_idx = nan_age[nan_age].index
nan_idx

Int64Index([  5,  17,  19,  26,  28,  29,  31,  32,  36,  42,
            ...
            832, 837, 839, 846, 849, 859, 863, 868, 878, 888],
           dtype='int64', length=177)

In [87]:
fill_med = lambda g: g.fillna(g.median())
df_ttt = df_ttt.groupby(['pclass', 'sex']).apply(fill_med)
df_ttt.info()
set(df_ttt.loc[nan_idx, 'age'])

<class 'pandas.core.frame.DataFrame'>
Int64Index: 891 entries, 0 to 890
Data columns (total 8 columns):
 #   Column    Non-Null Count  Dtype   
---  ------    --------------  -----   
 0   survived  891 non-null    int64   
 1   pclass    891 non-null    int64   
 2   sex       891 non-null    object  
 3   age       891 non-null    float64 
 4   fare      891 non-null    float64 
 5   embarked  889 non-null    object  
 6   deck      203 non-null    category
 7   family    891 non-null    int64   
dtypes: category(1), float64(2), int64(3), object(2)
memory usage: 89.2+ KB


{21.5, 25.0, 28.0, 30.0, 35.0, 40.0}

### deck 삭제 여부 판단

- deck (1/2/3 등급이 각각 -E/ D-F/ D-G 사용) 객실이 위치한 갑판
- nan은 해당 승객이 배정받은 객실 자체가 없어서 nan으로 설정 되었나 싶기도 하다. 입석같이. 
- deck에 따른 fare/ survived 확인 필요.

In [88]:
g_deck = df_ttt.groupby('deck')
list(g_deck)

[('A',
       survived  pclass     sex   age     fare embarked deck  family
  23          1       1    male  28.0  35.5000        S    A       0
  96          0       1    male  71.0  34.6542        C    A       0
  174         0       1    male  56.0  30.6958        C    A       0
  185         0       1    male  40.0  50.0000        S    A       0
  209         1       1    male  40.0  31.0000        C    A       0
  284         0       1    male  40.0  26.0000        S    A       0
  445         1       1    male   4.0  81.8583        S    A       2
  475         0       1    male  40.0  52.0000        S    A       0
  556         1       1  female  48.0  39.6000        C    A       1
  583         0       1    male  36.0  40.1250        C    A       0
  599         1       1    male  49.0  56.9292        C    A       1
  630         1       1    male  80.0  30.0000        S    A       0
  647         1       1    male  56.0  35.5000        C    A       0
  806         0       1    

In [89]:
df_ttt['deck'].info()
sr_nan = df_ttt[df_ttt['deck'].isna()]
sr_nan.head()

<class 'pandas.core.series.Series'>
Int64Index: 891 entries, 0 to 890
Series name: deck
Non-Null Count  Dtype   
--------------  -----   
203 non-null    category
dtypes: category(1)
memory usage: 40.5 KB


survived  pclass     sex   age     fare embarked deck  family
0         0       3    male  22.0   7.2500        S  NaN       1
2         1       3  female  26.0   7.9250        S  NaN       0
4         0       3    male  35.0   8.0500        S  NaN       0
5         0       3    male  25.0   8.4583        Q  NaN       0
7         0       3    male   2.0  21.0750        S  NaN       4

In [90]:
g_deck.mean()

survived    pclass        age        fare    family
deck                                                     
A     0.466667  1.000000  43.866667   39.623887  0.266667
B     0.744681  1.000000  35.063830  113.505764  0.936170
C     0.593220  1.000000  36.447797  100.151341  1.118644
D     0.757576  1.121212  38.939394   57.244576  0.727273
E     0.750000  1.312500  37.703125   46.026694  0.625000
F     0.615385  2.384615  20.461538   18.696792  1.076923
G     0.500000  3.000000  14.750000   13.581250  1.750000

In [91]:
sr_nan.mean()

survived     0.299419
pclass       2.636628
age         27.131177
fare        19.181079
family       0.911337
dtype: float64

# apply vs. applymap vs. pipe  

<img src = 'dataset/apply_applymap_pipe.jpg' STYLE='width:800px;'>
[pipe vs applymap vs apply]  

pipe와 applymap과의 차이는 반환하는 결과가 pipe가 더 다양하다는 것으로 applymap은 데이터프레임 원소에 적용하므로 데이터프레임으로만 리턴되지만 pipe는 사용자 함수의 리턴값에 따라 개별 값, 시리즈, 데이터프레임을 반환할 수 있다.  

현재 가장 많이 쓰이는 메서드는 apply인데 그 이유는 다양한 형태로 값을 입력 받을 수 있고, 반환하는 형태도 폭이 넓기 때문. 결과는 같아보이지만, 시리즈를 이용해 데이터프레임의 행과 열도 이용할 수 있으며 시리즈 그 자체도 함수로 활용할 수 있다. 

만약 데이터프레임에 함수를 적용해야한다면 리턴하는 형태가 많은 pipe 사용도 고려


## apply

In [92]:
import pandas as pd
df = pd.DataFrame({
    'Col1': [30, 40, 50, 60],
    'Col2': [23, 35, 65, 45],
    'Col3': [85, 87, 90, 89],
}, index=list('ABCD'))
df

Col1  Col2  Col3
A    30    23    85
B    40    35    87
C    50    65    90
D    60    45    89

In [93]:
df.apply(lambda x : x.sum(), axis=0) #x 는 df

Col1    180
Col2    168
Col3    351
dtype: int64

In [94]:
df.Col1.apply(lambda x : x+1)

A    31
B    41
C    51
D    61
Name: Col1, dtype: int64

In [95]:
df.Col1.map(lambda x : x+1)

A    31
B    41
C    51
D    61
Name: Col1, dtype: int64

## applymap

In [96]:
df.applymap(lambda x :'%.2f'%x)

Col1   Col2   Col3
A  30.00  23.00  85.00
B  40.00  35.00  87.00
C  50.00  65.00  90.00
D  60.00  45.00  89.00

## pipe

In [97]:
# The pipe() method allows you to apply one or more functions to the DataFrame object.

import pandas as pd

def change_age(x):
  x["age"]=[10, 20, 30]
  return x

data = {
  "name": ["Sally", "Mary", "John"],
  "age": [50, 40, 30]
}

df = pd.DataFrame(data)
df


name  age
0  Sally   50
1   Mary   40
2   John   30

In [98]:
df.pipe(change_age) # return도 하고 change도 하네

name  age
0  Sally   10
1   Mary   20
2   John   30

In [99]:
df

name  age
0  Sally   10
1   Mary   20
2   John   30

# 중복처리

In [100]:
df = pd.DataFrame({
    'c1':['a', 'a', 'b', 'a', 'b' ],
    'c2': [1,1,1,2,2],
    'c3': [1, 1, 2, 2, 2]
})
df

c1  c2  c3
0  a   1   1
1  a   1   1
2  b   1   2
3  a   2   2
4  b   2   2

## 중복 확인

In [101]:
# 전체 행 데이터중에서 중복
df_dup = df.duplicated()
df_dup

0    False
1     True
2    False
3    False
4    False
dtype: bool

In [102]:
#특정 열 데이터 중에서 중복
col_dup = df.c2.duplicated()
col_dup

0    False
1     True
2     True
3    False
4     True
Name: c2, dtype: bool

In [103]:
# 중복 행 제거
df2 = df.drop_duplicates()
df2

c1  c2  c3
0  a   1   1
2  b   1   2
3  a   2   2
4  b   2   2

In [104]:
df3 = df.drop_duplicates(subset=['c2','c3']) # columns의 list 넣기
df3

c1  c2  c3
0  a   1   1
2  b   1   2
3  a   2   2

# 데이터 단위 변경

In [105]:
auto_df = pd.read_csv('./dataset/auto-mpg.csv')
auto_df.head()

mpg  cylinders  displacement horsepower  weight  acceleration  model year  \
0  18.0          8         307.0        130    3504          12.0          70   
1  15.0          8         350.0        165    3693          11.5          70   
2  18.0          8         318.0        150    3436          11.0          70   
3  16.0          8         304.0        150    3433          12.0          70   
4  17.0          8         302.0        140    3449          10.5          70   

   origin                   car name  
0       1  chevrolet chevelle malibu  
1       1          buick skylark 320  
2       1         plymouth satellite  
3       1              amc rebel sst  
4       1                ford torino

In [106]:
# mpg - > kpl
mpg_to_kpl = 1.60934/3.78541
mpg_to_kpl

0.42514285110463595

In [107]:
auto_df['kpl'] = round(auto_df.mpg*mpg_to_kpl)#, 2)
auto_df[['mpg', 'kpl']].head()

mpg  kpl
0  18.0  8.0
1  15.0  6.0
2  18.0  8.0
3  16.0  7.0
4  17.0  7.0

# 데이터 타입 변경

In [108]:
auto_df.horsepower.unique()

array(['130', '165', '150', '140', '198', '220', '215', '225', '190',
       '170', '160', '95', '97', '85', '88', '46', '87', '90', '113',
       '200', '210', '193', '?', '100', '105', '175', '153', '180', '110',
       '72', '86', '70', '76', '65', '69', '60', '80', '54', '208', '155',
       '112', '92', '145', '137', '158', '167', '94', '107', '230', '49',
       '75', '91', '122', '67', '83', '78', '52', '61', '93', '148',
       '129', '96', '71', '98', '115', '53', '81', '79', '120', '152',
       '102', '108', '68', '58', '149', '89', '63', '48', '66', '139',
       '103', '125', '133', '138', '135', '142', '77', '62', '132', '84',
       '64', '74', '116', '82'], dtype=object)

In [109]:
import numpy as np
auto_df.horsepower.replace('?', np.nan, inplace=True)
auto_df.dropna(subset=['horsepower'], axis=0, inplace=True)
auto_df.horsepower.isnull().sum()

0

In [110]:
auto_df.horsepower.dtype
auto_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 392 entries, 0 to 397
Data columns (total 10 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   mpg           392 non-null    float64
 1   cylinders     392 non-null    int64  
 2   displacement  392 non-null    float64
 3   horsepower    392 non-null    object 
 4   weight        392 non-null    int64  
 5   acceleration  392 non-null    float64
 6   model year    392 non-null    int64  
 7   origin        392 non-null    int64  
 8   car name      392 non-null    object 
 9   kpl           392 non-null    float64
dtypes: float64(4), int64(4), object(2)
memory usage: 33.7+ KB


In [111]:
auto_df.horsepower = auto_df.horsepower.astype('int')
auto_df.horsepower.dtype

dtype('int32')

## (Done) 과제 #4 Fri 0701
 - 

'?'가 있는 인덱스(불린 인덱싱)를 찾아서 그 행을 삭제하세요.

In [112]:
df_car = pd.read_csv('./dataset/auto-mpg.csv')
df_car.head()

mpg  cylinders  displacement horsepower  weight  acceleration  model year  \
0  18.0          8         307.0        130    3504          12.0          70   
1  15.0          8         350.0        165    3693          11.5          70   
2  18.0          8         318.0        150    3436          11.0          70   
3  16.0          8         304.0        150    3433          12.0          70   
4  17.0          8         302.0        140    3449          10.5          70   

   origin                   car name  
0       1  chevrolet chevelle malibu  
1       1          buick skylark 320  
2       1         plymouth satellite  
3       1              amc rebel sst  
4       1                ford torino

In [113]:
df_car.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 398 entries, 0 to 397
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   mpg           398 non-null    float64
 1   cylinders     398 non-null    int64  
 2   displacement  398 non-null    float64
 3   horsepower    398 non-null    object 
 4   weight        398 non-null    int64  
 5   acceleration  398 non-null    float64
 6   model year    398 non-null    int64  
 7   origin        398 non-null    int64  
 8   car name      398 non-null    object 
dtypes: float64(3), int64(4), object(2)
memory usage: 28.1+ KB


In [114]:
df_car.horsepower.values

array(['130', '165', '150', '150', '140', '198', '220', '215', '225',
       '190', '170', '160', '150', '225', '95', '95', '97', '85', '88',
       '46', '87', '90', '95', '113', '90', '215', '200', '210', '193',
       '88', '90', '95', '?', '100', '105', '100', '88', '100', '165',
       '175', '153', '150', '180', '170', '175', '110', '72', '100', '88',
       '86', '90', '70', '76', '65', '69', '60', '70', '95', '80', '54',
       '90', '86', '165', '175', '150', '153', '150', '208', '155', '160',
       '190', '97', '150', '130', '140', '150', '112', '76', '87', '69',
       '86', '92', '97', '80', '88', '175', '150', '145', '137', '150',
       '198', '150', '158', '150', '215', '225', '175', '105', '100',
       '100', '88', '95', '46', '150', '167', '170', '180', '100', '88',
       '72', '94', '90', '85', '107', '90', '145', '230', '49', '75',
       '91', '112', '150', '110', '122', '180', '95', '?', '100', '100',
       '67', '80', '65', '75', '100', '110', '105', '140', '1

In [115]:
type(df_car.horsepower == '?')

pandas.core.series.Series

In [116]:
sr_bool = df_car.horsepower == '?'
sr_bool.sum()

6

In [117]:
sr_bool.values

array([False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False,  True, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
        True, False,

In [118]:
q_index = df_car[sr_bool].index
q_index

Int64Index([32, 126, 330, 336, 354, 374], dtype='int64')

In [119]:
df_car.iloc[q_index]

mpg  cylinders  displacement horsepower  weight  acceleration  \
32   25.0          4          98.0          ?    2046          19.0   
126  21.0          6         200.0          ?    2875          17.0   
330  40.9          4          85.0          ?    1835          17.3   
336  23.6          4         140.0          ?    2905          14.3   
354  34.5          4         100.0          ?    2320          15.8   
374  23.0          4         151.0          ?    3035          20.5   

     model year  origin              car name  
32           71       1            ford pinto  
126          74       1         ford maverick  
330          80       2  renault lecar deluxe  
336          80       1    ford mustang cobra  
354          81       2           renault 18i  
374          82       1        amc concord dl

In [120]:
df_carr = df_car.copy()
df_carr.drop(index=q_index, inplace=True)

In [121]:
df_carr.horsepower.unique()

array(['130', '165', '150', '140', '198', '220', '215', '225', '190',
       '170', '160', '95', '97', '85', '88', '46', '87', '90', '113',
       '200', '210', '193', '100', '105', '175', '153', '180', '110',
       '72', '86', '70', '76', '65', '69', '60', '80', '54', '208', '155',
       '112', '92', '145', '137', '158', '167', '94', '107', '230', '49',
       '75', '91', '122', '67', '83', '78', '52', '61', '93', '148',
       '129', '96', '71', '98', '115', '53', '81', '79', '120', '152',
       '102', '108', '68', '58', '149', '89', '63', '48', '66', '139',
       '103', '125', '133', '138', '135', '142', '77', '62', '132', '84',
       '64', '74', '116', '82'], dtype=object)

# 데이터 값 변경

In [122]:
auto_df = pd.read_csv('./dataset/auto-mpg.csv')
auto_df.head()

mpg  cylinders  displacement horsepower  weight  acceleration  model year  \
0  18.0          8         307.0        130    3504          12.0          70   
1  15.0          8         350.0        165    3693          11.5          70   
2  18.0          8         318.0        150    3436          11.0          70   
3  16.0          8         304.0        150    3433          12.0          70   
4  17.0          8         302.0        140    3449          10.5          70   

   origin                   car name  
0       1  chevrolet chevelle malibu  
1       1          buick skylark 320  
2       1         plymouth satellite  
3       1              amc rebel sst  
4       1                ford torino

In [123]:
auto_df.origin.unique()

array([1, 3, 2], dtype=int64)

- Q. 아래 사항을 처리 하세요.
* origin 열의 고유값을 출력하세요.
* 정수형 데이터를 문자형 데이터로 변환한 후 고유값을 출력하세요.(1:'USA', 2:'EU', 3:'JAPAN')


In [124]:
auto_df.origin.unique()

array([1, 3, 2], dtype=int64)

In [125]:
auto_df.replace({1:'USA', 2:'EU', 3:'JAPAN'}, inplace=True)

In [126]:
auto_df.origin.head()

0    USA
1    USA
2    USA
3    USA
4    USA
Name: origin, dtype: object

# type 변경

In [127]:
auto_df.origin.unique()

array(['USA', 'JAPAN', 'EU'], dtype=object)

In [128]:
auto_df.origin.dtypes

dtype('O')

In [129]:
auto_df.origin = auto_df.origin.astype('category')
auto_df.origin.dtypes

CategoricalDtype(categories=['EU', 'JAPAN', 'USA'], ordered=False)

In [130]:
auto_df.origin = auto_df.origin.astype('str')
auto_df.origin.dtypes

dtype('O')

In [131]:
# 이것은 전처리 된 auto-mpg

In [132]:
auto_df  = pd.read_pickle('./dataset/auto-mpg.pkl')
auto_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 392 entries, 0 to 397
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   mpg           392 non-null    float64
 1   cylinders     392 non-null    int64  
 2   displacement  392 non-null    float64
 3   horsepower    392 non-null    float64
 4   weight        392 non-null    int64  
 5   acceleration  392 non-null    float64
 6   model year    392 non-null    int64  
 7   origin        392 non-null    int64  
 8   car name      392 non-null    object 
dtypes: float64(4), int64(4), object(1)
memory usage: 30.6+ KB


In [133]:
# horsepower를 3개의 bin으로 나누고 
# bin_names = ['저출력', '보통출력', '고출력']

# numpy.histogram으로 경계값 뽑아서

In [134]:
import numpy as np
count, bin_dividers = np.histogram(auto_df.horsepower, bins=3)
print(type(count), type(bin_dividers))
print(count, bin_dividers)

bin_names = ['저출력', '보통출력', '고출력']

auto_df['hp_bin'] =pd.cut(x = auto_df.horsepower, #데이터 배열
                         bins=bin_dividers, # 경계값 리스트
                         labels=bin_names, #bin 이름
                         include_lowest=True) #첫 경계값 포함
auto_df[['horsepower','hp_bin']].head()

<class 'numpy.ndarray'> <class 'numpy.ndarray'>
[257 103  32] [ 46.         107.33333333 168.66666667 230.        ]


horsepower hp_bin
0       130.0   보통출력
1       165.0   보통출력
2       150.0   보통출력
3       150.0   보통출력
4       140.0   보통출력

In [135]:
bin_dividers.shape

(4,)

In [136]:
print(bin_dividers[3]-bin_dividers[2], bin_dividers[2]-bin_dividers[1], bin_dividers[1]-bin_dividers[0])

61.333333333333314 61.33333333333334 61.33333333333334


In [137]:
# 각 구간은 앞의 값 포함, 뒤의 값 미포함, 다만 마지막 간격은 뒤의 값도 포함

In [248]:
print(count) # 누적구간 

[257 103  32]


In [252]:
print(auto_df.horsepower)

0     0.5652
1     0.7174
2     0.6522
3     0.6522
4     0.6087
       ...  
393   0.3739
394   0.2261
395   0.3652
396   0.3435
397   0.3565
Name: horsepower, Length: 392, dtype: float64


In [253]:
print(bin_dividers) #간격 같음

[ 46.         107.33333333 168.66666667 230.        ]


In [140]:
auto_df.horsepower.min(), auto_df.horsepower.max()

(46.0, 230.0)

In [141]:
auto_df.horsepower

0      130.0
1      165.0
2      150.0
3      150.0
4      140.0
       ...  
393     86.0
394     52.0
395     84.0
396     79.0
397     82.0
Name: horsepower, Length: 392, dtype: float64

In [142]:
# Q. horsepower 열의 최대값의 절대값으로 모든 데이터를 
#나누는 정규화를 수행한 후 통계요약표를 출력하세요.

In [143]:
a_max = abs(auto_df.horsepower.max())

In [254]:
auto_df.horsepower= auto_df.horsepower/a_max
auto_df[['horsepower']].describe()

horsepower
count    392.0000
mean       0.0020
std        0.0007
min        0.0009
25%        0.0014
50%        0.0018
75%        0.0024
max        0.0043

In [257]:
fifa = pd.read_csv('./dataset/FIFA.csv', encoding='cp949')
fifa.head()

ID               Name  Age Nationality  Overall                 Club  \
0  158023           L. Messi   31   Argentina       94         FC Barcelona   
1   20801  Cristiano Ronaldo   33    Portugal       94             Juventus   
2  190871          Neymar Jr   26      Brazil       92  Paris Saint-Germain   
3  193080             De Gea   27       Spain       91    Manchester United   
4  192985       K. De Bruyne   27     Belgium       91      Manchester City   

  Preferred_Foot       Work_Rate Position  Jersey_Number  \
0           Left  Medium/ Medium       RF             10   
1          Right       High/ Low       ST              7   
2          Right    High/ Medium       LW             10   
3          Right  Medium/ Medium       GK              1   
4          Right      High/ High      RCM              7   

   Contract_Valid_Until Height  Weight_lb  Release_Clause   Value  Wage  
0                  2021    5'7        159          226500  110500   565  
1                  2022    6'2        183          127100   77000   405  
2                  2022    5'9        150          228100  118500   290  
3                  2020    6'4        168          138600   72000   260  
4                  2023   5'11        154          196400  102000   355

## (Done) 과제 #1 Mon 0704
Q. Height 값을 cm로 변환하여 컬럼 'height_cm'을 생성

In [259]:
fifa.Height, type(fifa.Height)

(0         5'7
 1         6'2
 2         5'9
 3         6'4
 4        5'11
          ... 
 16637     5'9
 16638     6'3
 16639     5'8
 16640    5'10
 16641    5'10
 Name: Height, Length: 16642, dtype: object,
 pandas.core.series.Series)

In [260]:
df = fifa.copy()
df.head()

ID               Name  Age Nationality  Overall                 Club  \
0  158023           L. Messi   31   Argentina       94         FC Barcelona   
1   20801  Cristiano Ronaldo   33    Portugal       94             Juventus   
2  190871          Neymar Jr   26      Brazil       92  Paris Saint-Germain   
3  193080             De Gea   27       Spain       91    Manchester United   
4  192985       K. De Bruyne   27     Belgium       91      Manchester City   

  Preferred_Foot       Work_Rate Position  Jersey_Number  \
0           Left  Medium/ Medium       RF             10   
1          Right       High/ Low       ST              7   
2          Right    High/ Medium       LW             10   
3          Right  Medium/ Medium       GK              1   
4          Right      High/ High      RCM              7   

   Contract_Valid_Until Height  Weight_lb  Release_Clause   Value  Wage  
0                  2021    5'7        159          226500  110500   565  
1                  2022    6'2        183          127100   77000   405  
2                  2022    5'9        150          228100  118500   290  
3                  2020    6'4        168          138600   72000   260  
4                  2023   5'11        154          196400  102000   355

In [261]:
def fech_to_cm(x):
    x.split("'")

In [262]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16642 entries, 0 to 16641
Data columns (total 16 columns):
 #   Column                Non-Null Count  Dtype 
---  ------                --------------  ----- 
 0   ID                    16642 non-null  int64 
 1   Name                  16642 non-null  object
 2   Age                   16642 non-null  int64 
 3   Nationality           16642 non-null  object
 4   Overall               16642 non-null  int64 
 5   Club                  16642 non-null  object
 6   Preferred_Foot        16642 non-null  object
 7   Work_Rate             16642 non-null  object
 8   Position              16642 non-null  object
 9   Jersey_Number         16642 non-null  int64 
 10  Contract_Valid_Until  16642 non-null  int64 
 11  Height                16642 non-null  object
 12  Weight_lb             16642 non-null  int64 
 13  Release_Clause        16642 non-null  int64 
 14  Value                 16642 non-null  int64 
 15  Wage                  16642 non-null

In [263]:
def fech_to_cm(x):
    feet = int(x.split( "'")[0])
    inch= int(x.split( "'")[1])
    return feet*30.48 + inch*2.54

In [264]:
df['height_cm'] = df.Height.apply(fech_to_cm)

In [265]:
df.head()

ID               Name  Age Nationality  Overall                 Club  \
0  158023           L. Messi   31   Argentina       94         FC Barcelona   
1   20801  Cristiano Ronaldo   33    Portugal       94             Juventus   
2  190871          Neymar Jr   26      Brazil       92  Paris Saint-Germain   
3  193080             De Gea   27       Spain       91    Manchester United   
4  192985       K. De Bruyne   27     Belgium       91      Manchester City   

  Preferred_Foot       Work_Rate Position  Jersey_Number  \
0           Left  Medium/ Medium       RF             10   
1          Right       High/ Low       ST              7   
2          Right    High/ Medium       LW             10   
3          Right  Medium/ Medium       GK              1   
4          Right      High/ High      RCM              7   

   Contract_Valid_Until Height  Weight_lb  Release_Clause   Value  Wage  \
0                  2021    5'7        159          226500  110500   565   
1                  2022    6'2        183          127100   77000   405   
2                  2022    5'9        150          228100  118500   290   
3                  2020    6'4        168          138600   72000   260   
4                  2023   5'11        154          196400  102000   355   

   height_cm  
0   170.1800  
1   187.9600  
2   175.2600  
3   193.0400  
4   180.3400

# 데이터프레임 결합 join
* 판다스 join() 메소드는 merge() 함수를 기반으로 만들어졌기 때문에 기본 작동 방식이 서로 비슷
* join() 메소드는 두 데이터프레임의 행 인덱스를 기준으로 결합하는 점에서 merge와 차이가 있으나 on=keys 옵션 설정하면 열을 기준으로 결합


In [269]:
data1 = np.arange(1, 21).reshape(4, 5)
data2 = np.arange(11, 31).reshape(5, 4)
df1 = pd.DataFrame(data1, columns=list('abcde'))
df2 = pd.DataFrame(data2, columns=list('dehi'))

In [270]:
print(df1, '\n')
print(df2)

    a   b   c   d   e
0   1   2   3   4   5
1   6   7   8   9  10
2  11  12  13  14  15
3  16  17  18  19  20 

    d   e   h   i
0  11  12  13  14
1  15  16  17  18
2  19  20  21  22
3  23  24  25  26
4  27  28  29  30


In [271]:
df1.join(df2, lsuffix='_a', rsuffix='_b')

a   b   c  d_a  e_a  d_b  e_b   h   i
0   1   2   3    4    5   11   12  13  14
1   6   7   8    9   10   15   16  17  18
2  11  12  13   14   15   19   20  21  22
3  16  17  18   19   20   23   24  25  26

In [272]:
df1.join(df2, lsuffix='_a', rsuffix='_b', how = 'inner')

a   b   c  d_a  e_a  d_b  e_b   h   i
0   1   2   3    4    5   11   12  13  14
1   6   7   8    9   10   15   16  17  18
2  11  12  13   14   15   19   20  21  22
3  16  17  18   19   20   23   24  25  26

In [273]:
df1.join(df2, lsuffix='_a', rsuffix='_b', how = 'outer')

a       b       c     d_a     e_a  d_b  e_b   h   i
0  1.0000  2.0000  3.0000  4.0000  5.0000   11   12  13  14
1  6.0000  7.0000  8.0000  9.0000 10.0000   15   16  17  18
2 11.0000 12.0000 13.0000 14.0000 15.0000   19   20  21  22
3 16.0000 17.0000 18.0000 19.0000 20.0000   23   24  25  26
4     NaN     NaN     NaN     NaN     NaN   27   28  29  30

In [276]:
df1['ind'] = [1, 2, 3, 4]
df2['ind'] = [1, 2, 3, 4, 5]
df1

a   b   c   d   e  ind
0   1   2   3   4   5    1
1   6   7   8   9  10    2
2  11  12  13  14  15    3
3  16  17  18  19  20    4

In [277]:
df2

d   e   h   i  ind
0  11  12  13  14    1
1  15  16  17  18    2
2  19  20  21  22    3
3  23  24  25  26    4
4  27  28  29  30    5

In [159]:
df1.set_index('ind').join(df2.set_index('ind'), lsuffix='_a', rsuffix='_b')

a   b   c  d_a  e_a  d_b  e_b   h   i
ind                                        
1     1   2   3    4    5   11   12  13  14
2     6   7   8    9   10   15   16  17  18
3    11  12  13   14   15   19   20  21  22
4    16  17  18   19   20   23   24  25  26

In [278]:
df1 # set_index는 inplace 옵션이 있다, 하면 어디 붙음?

a   b   c   d   e  ind
0   1   2   3   4   5    1
1   6   7   8   9  10    2
2  11  12  13  14  15    3
3  16  17  18  19  20    4

In [279]:
df1.set_index('ind', inplace=True)
df1

a   b   c   d   e
ind                    
1     1   2   3   4   5
2     6   7   8   9  10
3    11  12  13  14  15
4    16  17  18  19  20

## join options

In [161]:
!pip install openpyxl

In [280]:
df_s1 = pd.read_excel('./dataset/stock price.xlsx', 
                      engine='openpyxl', index_col = 'id')
df_s2 = pd.read_excel('./dataset/stock valuation.xlsx', 
                      engine='openpyxl', index_col = 'id')
df_s1.head()

stock_name       value   price
id                                   
128940       한미약품  59385.6667  421000
130960     CJ E&M  58540.6667   98900
138250      엔에스쇼핑  14558.6667   13200
139480        이마트 239230.8333  254500
142280     녹십자엠에스    468.8333   10200

In [281]:
df_s2.head()

name        eps     bps     per    pbr
id                                               
130960   CJ E&M  6301.3333   54068 15.6951 1.8292
136480       하림   274.1667    3551 11.4894 0.8871
138040  메리츠금융지주  2122.3333   14894  6.3138 0.8997
139480      이마트 18268.1667  295780 13.9313 0.8604
145990      삼양사  5741.0000  108090 14.2832 0.7586

In [164]:
df_j1 = df_s1.join(df_s2, how = 'right')
df_j1.head()

stock_name          value     price     name           eps     bps  \
id                                                                          
130960     CJ E&M   58540.666667   98900.0   CJ E&M   6301.333333   54068   
136480        NaN            NaN       NaN       하림    274.166667    3551   
138040        NaN            NaN       NaN  메리츠금융지주   2122.333333   14894   
139480        이마트  239230.833333  254500.0      이마트  18268.166667  295780   
145990        삼양사   82750.000000   82000.0      삼양사   5741.000000  108090   

              per       pbr  
id                           
130960  15.695091  1.829178  
136480  11.489362  0.887074  
138040   6.313806  0.899691  
139480  13.931338  0.860437  
145990  14.283226  0.758627

In [165]:
df_j2 = df_s1.join(df_s2, how = 'left')
df_j2.head()

stock_name          value   price    name           eps       bps  \
id                                                                         
128940       한미약품   59385.666667  421000     NaN           NaN       NaN   
130960     CJ E&M   58540.666667   98900  CJ E&M   6301.333333   54068.0   
138250      엔에스쇼핑   14558.666667   13200     NaN           NaN       NaN   
139480        이마트  239230.833333  254500     이마트  18268.166667  295780.0   
142280     녹십자엠에스     468.833333   10200     NaN           NaN       NaN   

              per       pbr  
id                           
128940        NaN       NaN  
130960  15.695091  1.829178  
138250        NaN       NaN  
139480  13.931338  0.860437  
142280        NaN       NaN

In [166]:
df_j2 = df_s1.join(df_s2, how = 'inner')
df_j2.head()

stock_name          value   price    name           eps     bps  \
id                                                                       
130960     CJ E&M   58540.666667   98900  CJ E&M   6301.333333   54068   
139480        이마트  239230.833333  254500     이마트  18268.166667  295780   
145990        삼양사   82750.000000   82000     삼양사   5741.000000  108090   
185750        종근당   40293.666667  100500     종근당   3990.333333   40684   
204210     모두투어리츠    3093.333333    3475  모두투어리츠     85.166667    5335   

              per       pbr  
id                           
130960  15.695091  1.829178  
139480  13.931338  0.860437  
145990  14.283226  0.758627  
185750  25.185866  2.470259  
204210  40.802348  0.651359

In [ ]:
df = pd.DataFrame({'key': ['K0', 'K1', 'K2', 'K3', 'K4', 'K5'],
                   'A': ['A0', 'A1', 'A2', 'A3', 'A4', 'A5']})

## (Done) 과제 #2 Mon 0704
Q. df_j2에서 중복된 칼럼을 삭제한 후 블린 인덱싱을 이용하여 eps가 3000 보다 적거나 stock_name이 이마트인 데이터를 선택하여 
데이터프레임을 생성하고 df_j3 이름으로 저장 및 출력하세요.(단, '<' 와 '==' 를 반드시 사용해야 함)


In [167]:
df_j2.drop(columns='name', inplace=True)
df_j2

stock_name          value   price           eps     bps        per  \
id                                                                          
130960     CJ E&M   58540.666667   98900   6301.333333   54068  15.695091   
139480        이마트  239230.833333  254500  18268.166667  295780  13.931338   
145990        삼양사   82750.000000   82000   5741.000000  108090  14.283226   
185750        종근당   40293.666667  100500   3990.333333   40684  25.185866   
204210     모두투어리츠    3093.333333    3475     85.166667    5335  40.802348   

             pbr  
id                
130960  1.829178  
139480  0.860437  
145990  0.758627  
185750  2.470259  
204210  0.651359

In [168]:
df_j3 = df_j2[(df_j2.eps < 3000) | (df_j2.stock_name == '이마트')]
df_j3

stock_name          value   price           eps     bps        per  \
id                                                                          
139480        이마트  239230.833333  254500  18268.166667  295780  13.931338   
204210     모두투어리츠    3093.333333    3475     85.166667    5335  40.802348   

             pbr  
id                
139480  0.860437  
204210  0.651359

## ★JOIN 테스트

In [287]:
df = pd.DataFrame({'key': ['K0', 'K1', 'K2', 'K3', 'K4', 'K5'],
                   'A': ['A0', 'A1', 'A2', 'A3', 'A4', 'A5']})
df

key   A
0  K0  A0
1  K1  A1
2  K2  A2
3  K3  A3
4  K4  A4
5  K5  A5

In [292]:
other = pd.DataFrame({'key': ['K0', 'K2', 'K3'],
                      'B': ['B0', 'B1', 'B2']})
other

key   B
0  K0  B0
1  K2  B1
2  K3  B2

In [293]:
df.join(other, lsuffix='_caller', rsuffix='_other')

key_caller   A key_other    B
0         K0  A0        K0   B0
1         K1  A1        K2   B1
2         K2  A2        K3   B2
3         K3  A3       NaN  NaN
4         K4  A4       NaN  NaN
5         K5  A5       NaN  NaN

In [294]:
df.set_index('key').join(other.set_index('key'))

A    B
key         
K0   A0   B0
K1   A1  NaN
K2   A2   B1
K3   A3   B2
K4   A4  NaN
K5   A5  NaN

In [295]:
df.join(other.set_index('key'), on='key')

key   A    B
0  K0  A0   B0
1  K1  A1  NaN
2  K2  A2   B1
3  K3  A3   B2
4  K4  A4  NaN
5  K5  A5  NaN

In [296]:
df = pd.DataFrame({'key': ['K0', 'K1', 'K1', 'K3', 'K0', 'K1'],
                   'A': ['A0', 'A1', 'A2', 'A3', 'A4', 'A5']})
df

key   A
0  K0  A0
1  K1  A1
2  K1  A2
3  K3  A3
4  K0  A4
5  K1  A5

In [298]:
other

key   B
0  K0  B0
1  K2  B1
2  K3  B2

In [299]:
df.join(other.set_index('key'), on='key')

key   A    B
0  K0  A0   B0
1  K1  A1  NaN
2  K1  A2  NaN
3  K3  A3   B2
4  K0  A4   B0
5  K1  A5  NaN

# 데이터프레임 병합 merge
* merge() 함수는 sql의 join 명령과 비슷한 방식으로 어떤 기준에 의해 두 데이터프레임을 병합한는 개념
* 기준이 되는 열이나 인덱스를 키라고 부름
* 키가 되는 열이나 인덱스는 반드시 양쪽 데이터프레임에 모두 존재함


In [300]:
data1 = np.arange(1, 21).reshape(4, 5)
data2 = np.arange(11, 31).reshape(5, 4)
df1 = pd.DataFrame(data1, columns=list('abcde'))
df2 = pd.DataFrame(data2, columns=list('dehi'))
print(df1, '\n')
print(df2)

    a   b   c   d   e
0   1   2   3   4   5
1   6   7   8   9  10
2  11  12  13  14  15
3  16  17  18  19  20 

    d   e   h   i
0  11  12  13  14
1  15  16  17  18
2  19  20  21  22
3  23  24  25  26
4  27  28  29  30


In [301]:
df1['ind'] = [1, 2, 3, 4]
df2['ind'] = [1, 2, 3, 4, 5]
df1

a   b   c   d   e  ind
0   1   2   3   4   5    1
1   6   7   8   9  10    2
2  11  12  13  14  15    3
3  16  17  18  19  20    4

In [302]:
df2

d   e   h   i  ind
0  11  12  13  14    1
1  15  16  17  18    2
2  19  20  21  22    3
3  23  24  25  26    4
4  27  28  29  30    5

In [303]:
pd.merge(df1, df2, on='ind', how = 'outer')

a       b       c     d_x     e_x  ind  d_y  e_y   h   i
0  1.0000  2.0000  3.0000  4.0000  5.0000    1   11   12  13  14
1  6.0000  7.0000  8.0000  9.0000 10.0000    2   15   16  17  18
2 11.0000 12.0000 13.0000 14.0000 15.0000    3   19   20  21  22
3 16.0000 17.0000 18.0000 19.0000 20.0000    4   23   24  25  26
4     NaN     NaN     NaN     NaN     NaN    5   27   28  29  30

In [304]:
df1.count()

a      4
b      4
c      4
d      4
e      4
ind    4
dtype: int64

## ★MERGE 테스트

In [337]:
df1 = pd.DataFrame({'lkey': ['foo', 'bar', 'baz', 'foo'],
                    'value': [1, 2, 3, 5]})
df2 = pd.DataFrame({'rkey': ['foo', 'bar', 'baz', 'foo'],
                    'value': [5, 6, 7, 8]})
df1

lkey  value
0  foo      1
1  bar      2
2  baz      3
3  foo      5

In [338]:
df2

rkey  value
0  foo      5
1  bar      6
2  baz      7
3  foo      8

In [339]:
df1.merge(df2, left_on='lkey', right_on='rkey')

lkey  value_x rkey  value_y
0  foo        1  foo        5
1  foo        1  foo        8
2  foo        5  foo        5
3  foo        5  foo        8
4  bar        2  bar        6
5  baz        3  baz        7

In [342]:
df1

lkey  value
0  foo      1
1  bar      2
2  baz      3
3  foo      5

In [343]:
df2

rkey  value
0  foo      5
1  bar      6
2  baz      7
3  foo      8

In [379]:
df1 = pd.DataFrame({'a': ['fo', 'bar'], 'b': [1, 2]})
df2 = pd.DataFrame({'a': ['foo', 'baz'], 'c': [3, 4]})
df1

a  b
0   fo  1
1  bar  2

In [380]:
df2

a  c
0  foo  3
1  baz  4

In [381]:
df1.merge(df2, how='inner')

Empty DataFrame
Columns: [a, b, c]
Index: []

In [382]:
df1.merge(df2, how='outer')

a      b      c
0   fo 1.0000    NaN
1  bar 2.0000    NaN
2  foo    NaN 3.0000
3  baz    NaN 4.0000

In [383]:
df1.merge(df2, how='left')

a  b   c
0   fo  1 NaN
1  bar  2 NaN

In [384]:
df1.merge(df2, how='right')

a   b  c
0  foo NaN  3
1  baz NaN  4

In [365]:
df1.merge(df2, how='inner', on='a')

a  b  c
0  foo  1  3

In [366]:
df1.merge(df2, how='left', on='a') # 왼쪽 value들은 남김.

a  b      c
0  foo  1 3.0000
1  bar  2    NaN

In [369]:
df1.merge(df2, how='right', on='a') # 오른쪽 value들은 남김.

a      b  c
0  foo 1.0000  3
1  baz    NaN  4

In [359]:
df1 = pd.DataFrame({'left': ['foo', 'bar']})
df2 = pd.DataFrame({'right': [7, 8]})
df1

left
0  foo
1  bar

In [360]:
df2

right
0      7
1      8

In [353]:
df1.merge(df2, how='cross')

left  right
0  foo      7
1  foo      8
2  bar      7
3  bar      8

In [330]:
df1.merge(df2, left_on='lkey', right_on='rkey',
          suffixes=('_left', '_right'))

lkey  value_left rkey  value_right
0  foo           1  foo            5
1  foo           1  foo            8
2  foo           5  foo            5
3  foo           5  foo            8
4  bar           2  bar            6
5  baz           3  baz            7

In [331]:
df1 = pd.DataFrame({'a': ['foo', 'bar'], 'b': [1, 2]})
df2 = pd.DataFrame({'a': ['foo', 'baz'], 'c': [3, 4]})
df1.merge(df2, how='inner', on='a')

a  b  c
0  foo  1  3

In [332]:
df1.merge(df2, how='left', on='a')

a  b      c
0  foo  1 3.0000
1  bar  2    NaN

## (Done) 과제 #3 Mon 0704

- on='None'으로 교집합을 만들어 출력하세요.

- 첫번째 df는(stock_name), 두번째 df는 (name)을 기준으로 합집합(left)을 만들어 출력하세요.

- 오른쪽 df를 기준으로 합친 후 출력하세요

- df1에서 가격이 50000 미만인 데이터만 출력하세요

- df2에서 per가 30보다 크고 pbr이 0.8보다 큰 데이터를 모두 출력하세요

In [385]:
df1 = pd.read_csv('./dataset/stock_price.csv', encoding='cp949')
df1

id stock_name       value   price
0  128940       한미약품  59385.6667  421000
1  130960     CJ E&M  58540.6667   98900
2  138250      엔에스쇼핑  14558.6667   13200
3  139480        이마트 239230.8333  254500
4  142280     녹십자엠에스    468.8333   10200
5  145990        삼양사  82750.0000   82000
6  185750        종근당  40293.6667  100500
7  192400      쿠쿠홀딩스 179204.6667  177500
8  199800         툴젠  -2514.3333  115400
9  204210     모두투어리츠   3093.3333    3475

In [399]:
df2 = pd.read_csv('./dataset/stock_valuation.csv', encoding='cp949')
df2

id       name        eps     bps     per    pbr
0  130960     CJ E&M  6301.3333   54068 15.6951 1.8292
1  136480         하림   274.1667    3551 11.4894 0.8871
2  138040    메리츠금융지주  2122.3333   14894  6.3138 0.8997
3  139480        이마트 18268.1667  295780 13.9313 0.8604
4  145990        삼양사  5741.0000  108090 14.2832 0.7586
5  161390      한국타이어  5648.5000   51341  7.4533 0.8200
6  181710  NHN엔터테인먼트  2110.1667   78434 30.7559 0.8274
7  185750        종근당  3990.3333   40684 25.1859 2.4703
8  204210     모두투어리츠    85.1667    5335 40.8023 0.6514
9  207940   삼성바이오로직스  4644.1667   60099 89.7901 6.9386

df1, df2에 대하여 다음을 수행하세요.
- on='None'으로 교집합을 만들어 출력하세요.
- 첫번째 df는(stock_name), 두번째 df는 (name)을 기준으로 합집합(left)을 만들어 출력하세요.
- 오른쪽 df를 기준으로 합친 후 출력하세요
- df1에서 가격이 50000 미만인 데이터만 출력하세요
- df2에서 per가 30보다 크고 pbr이 0.8보다 큰 데이터를 모두 출력하세요

- on='None'으로 교집합을 만들어 출력하세요.

In [400]:
df1.merge(df2, how='inner', on=None) # 이게 뭔기준임

id stock_name       value   price    name        eps     bps     per  \
0  130960     CJ E&M  58540.6667   98900  CJ E&M  6301.3333   54068 15.6951   
1  139480        이마트 239230.8333  254500     이마트 18268.1667  295780 13.9313   
2  145990        삼양사  82750.0000   82000     삼양사  5741.0000  108090 14.2832   
3  185750        종근당  40293.6667  100500     종근당  3990.3333   40684 25.1859   
4  204210     모두투어리츠   3093.3333    3475  모두투어리츠    85.1667    5335 40.8023   

     pbr  
0 1.8292  
1 0.8604  
2 0.7586  
3 2.4703  
4 0.6514

- 첫번째 df는(stock_name), 두번째 df는 (name)을 기준으로 합집합(left)을 만들어 출력하세요.

In [401]:
df2.rename(columns={'name':'stock_name'}, inplace=True)

In [402]:
# merge의 left_on, right_on 있어서 rename 불필요

In [403]:
pd.merge(df1, df2, on='stock_name', how = 'left')

id_x stock_name       value   price        id_y        eps         bps  \
0  128940       한미약품  59385.6667  421000         NaN        NaN         NaN   
1  130960     CJ E&M  58540.6667   98900 130960.0000  6301.3333  54068.0000   
2  138250      엔에스쇼핑  14558.6667   13200         NaN        NaN         NaN   
3  139480        이마트 239230.8333  254500 139480.0000 18268.1667 295780.0000   
4  142280     녹십자엠에스    468.8333   10200         NaN        NaN         NaN   
5  145990        삼양사  82750.0000   82000 145990.0000  5741.0000 108090.0000   
6  185750        종근당  40293.6667  100500 185750.0000  3990.3333  40684.0000   
7  192400      쿠쿠홀딩스 179204.6667  177500         NaN        NaN         NaN   
8  199800         툴젠  -2514.3333  115400         NaN        NaN         NaN   
9  204210     모두투어리츠   3093.3333    3475 204210.0000    85.1667   5335.0000   

      per    pbr  
0     NaN    NaN  
1 15.6951 1.8292  
2     NaN    NaN  
3 13.9313 0.8604  
4     NaN    NaN  
5 14.2832 0.7586  
6 25.1859 2.4703  
7     NaN    NaN  
8     NaN    NaN  
9 40.8023 0.6514

- 오른쪽 df를 기준으로 합친 후 출력하세요

In [396]:
pd.merge(df1, df2, on='stock_name', how = 'right')

id_x stock_name       value       price    id_y        eps     bps  \
0 130960.0000     CJ E&M  58540.6667  98900.0000  130960  6301.3333   54068   
1         NaN         하림         NaN         NaN  136480   274.1667    3551   
2         NaN    메리츠금융지주         NaN         NaN  138040  2122.3333   14894   
3 139480.0000        이마트 239230.8333 254500.0000  139480 18268.1667  295780   
4 145990.0000        삼양사  82750.0000  82000.0000  145990  5741.0000  108090   
5         NaN      한국타이어         NaN         NaN  161390  5648.5000   51341   
6         NaN  NHN엔터테인먼트         NaN         NaN  181710  2110.1667   78434   
7 185750.0000        종근당  40293.6667 100500.0000  185750  3990.3333   40684   
8 204210.0000     모두투어리츠   3093.3333   3475.0000  204210    85.1667    5335   
9         NaN   삼성바이오로직스         NaN         NaN  207940  4644.1667   60099   

      per    pbr  
0 15.6951 1.8292  
1 11.4894 0.8871  
2  6.3138 0.8997  
3 13.9313 0.8604  
4 14.2832 0.7586  
5  7.4533 0.8200  
6 30.7559 0.8274  
7 25.1859 2.4703  
8 40.8023 0.6514  
9 89.7901 6.9386

- df1에서 가격이 50000 미만인 데이터만 출력하세요

In [397]:
df1[df1.price<50000]

id stock_name      value  price
2  138250      엔에스쇼핑 14558.6667  13200
4  142280     녹십자엠에스   468.8333  10200
9  204210     모두투어리츠  3093.3333   3475

- df2에서 per가 30보다 크고 pbr이 0.8보다 큰 데이터를 모두 출력하세요

In [398]:
df2[(df2.per > 30) & (df2.pbr >0.8)]

id stock_name       eps    bps     per    pbr
6  181710  NHN엔터테인먼트 2110.1667  78434 30.7559 0.8274
9  207940   삼성바이오로직스 4644.1667  60099 89.7901 6.9386

## (Done) 과제 #4 Mon 0704

### Q16. 함수 매핑을 이용하여 age에 각각 10을 곱하고 5를 더한 후 다시 더한 결과 값이 출력되게 하세요.


In [405]:
# (a * b) + (a + c)
# 데이터 가져오기
import seaborn as sns

# titanic 데이터셋에서 

titanic = sns.load_dataset('titanic')
# print(titanic.head())
df = titanic.loc[:, ['age','fare']]
print(df.head())


      age    fare
0 22.0000  7.2500
1 38.0000 71.2833
2 26.0000  7.9250
3 35.0000 53.1000
4 35.0000  8.0500


In [406]:
def f_muladd(x):
    return (x*10) + (x+5)

In [407]:
df.age.apply(f_muladd)

0     247.0000
1     423.0000
2     291.0000
3     390.0000
4     390.0000
        ...   
886   302.0000
887   214.0000
888        NaN
889   291.0000
890   357.0000
Name: age, Length: 891, dtype: float64

In [408]:
def f2_muladd(a, b, c):
    return (a*b)+(a+c)

In [409]:
df.age.apply(f2_muladd, b=10, c=5)

0     247.0000
1     423.0000
2     291.0000
3     390.0000
4     390.0000
        ...   
886   302.0000
887   214.0000
888        NaN
889   291.0000
890   357.0000
Name: age, Length: 891, dtype: float64

### Q17. titanic.fare 데이터를 소수 2째자리 형태로 모두 출력하세요.
- 7.2500 -> 7.25


In [410]:
df.fare.apply(lambda x :'%.2f'%x)

0       7.25
1      71.28
2       7.92
3      53.10
4       8.05
       ...  
886    13.00
887    30.00
888    23.45
889    30.00
890     7.75
Name: fare, Length: 891, dtype: object

In [411]:
df.fare.round(2)

0      7.2500
1     71.2800
2      7.9200
3     53.1000
4      8.0500
        ...  
886   13.0000
887   30.0000
888   23.4500
889   30.0000
890    7.7500
Name: fare, Length: 891, dtype: float64

### Q18. df에서 fare와 age를 선택하여 100을 곱한 후 출력하세요


In [412]:
import pandas as pd

titanic = sns.load_dataset('titanic')
df = titanic
df.head()


survived  pclass     sex     age  sibsp  parch    fare embarked  class  \
0         0       3    male 22.0000      1      0  7.2500        S  Third   
1         1       1  female 38.0000      1      0 71.2833        C  First   
2         1       3  female 26.0000      0      0  7.9250        S  Third   
3         1       1  female 35.0000      1      0 53.1000        S  First   
4         0       3    male 35.0000      0      0  8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True

In [413]:
df.fare*100

0      725.0000
1     7128.3300
2      792.5000
3     5310.0000
4      805.0000
         ...   
886   1300.0000
887   3000.0000
888   2345.0000
889   3000.0000
890    775.0000
Name: fare, Length: 891, dtype: float64

In [414]:
df.age*100

0     2200.0000
1     3800.0000
2     2600.0000
3     3500.0000
4     3500.0000
         ...   
886   2700.0000
887   1900.0000
888         NaN
889   2600.0000
890   3200.0000
Name: age, Length: 891, dtype: float64

### Q19. join() 메소드는 두 데이터프레임의 행 인덱스를 기준으로 결합한다. 2개의 주식데이터를 가져와서 join() 메소드로 아래와 같이 결합한 후 다음 사항을 수행하세요.
- df1과 df2의 교집합만 출력되도록 결합하여 df4에 저장하고 출력
- df4에서 중복된 칼럼을 삭제한 후 블린 인덱싱을 이용하여 eps가 3000 보다 적거나 stock_name이 이마트인 데이터를 선택하여 데이터프레임을 생성하고 df5 이름으로 저장 및 출력하세요.(단, '<' 와 '==' 를 반드시 사용해야 함)


In [415]:
df1 = pd.read_csv('./dataset/stock_price.csv', encoding='cp949')
df1

id stock_name       value   price
0  128940       한미약품  59385.6667  421000
1  130960     CJ E&M  58540.6667   98900
2  138250      엔에스쇼핑  14558.6667   13200
3  139480        이마트 239230.8333  254500
4  142280     녹십자엠에스    468.8333   10200
5  145990        삼양사  82750.0000   82000
6  185750        종근당  40293.6667  100500
7  192400      쿠쿠홀딩스 179204.6667  177500
8  199800         툴젠  -2514.3333  115400
9  204210     모두투어리츠   3093.3333    3475

In [416]:
df2 = pd.read_csv('./dataset/stock_valuation.csv', encoding='cp949')
df2

id       name        eps     bps     per    pbr
0  130960     CJ E&M  6301.3333   54068 15.6951 1.8292
1  136480         하림   274.1667    3551 11.4894 0.8871
2  138040    메리츠금융지주  2122.3333   14894  6.3138 0.8997
3  139480        이마트 18268.1667  295780 13.9313 0.8604
4  145990        삼양사  5741.0000  108090 14.2832 0.7586
5  161390      한국타이어  5648.5000   51341  7.4533 0.8200
6  181710  NHN엔터테인먼트  2110.1667   78434 30.7559 0.8274
7  185750        종근당  3990.3333   40684 25.1859 2.4703
8  204210     모두투어리츠    85.1667    5335 40.8023 0.6514
9  207940   삼성바이오로직스  4644.1667   60099 89.7901 6.9386

In [417]:
df4 = df1.join(df2, lsuffix='_a', rsuffix='_b', how='inner')
df4

id_a stock_name       value   price    id_b       name        eps  \
0  128940       한미약품  59385.6667  421000  130960     CJ E&M  6301.3333   
1  130960     CJ E&M  58540.6667   98900  136480         하림   274.1667   
2  138250      엔에스쇼핑  14558.6667   13200  138040    메리츠금융지주  2122.3333   
3  139480        이마트 239230.8333  254500  139480        이마트 18268.1667   
4  142280     녹십자엠에스    468.8333   10200  145990        삼양사  5741.0000   
5  145990        삼양사  82750.0000   82000  161390      한국타이어  5648.5000   
6  185750        종근당  40293.6667  100500  181710  NHN엔터테인먼트  2110.1667   
7  192400      쿠쿠홀딩스 179204.6667  177500  185750        종근당  3990.3333   
8  199800         툴젠  -2514.3333  115400  204210     모두투어리츠    85.1667   
9  204210     모두투어리츠   3093.3333    3475  207940   삼성바이오로직스  4644.1667   

      bps     per    pbr  
0   54068 15.6951 1.8292  
1    3551 11.4894 0.8871  
2   14894  6.3138 0.8997  
3  295780 13.9313 0.8604  
4  108090 14.2832 0.7586  
5   51341  7.4533 0.8200  
6   78434 30.7559 0.8274  
7   40684 25.1859 2.4703  
8    5335 40.8023 0.6514  
9   60099 89.7901 6.9386


- df4에서 중복된 칼럼을 삭제한 후 블린 인덱싱을 이용하여 eps가 3000 보다 적거나 stock_name이 이마트인 데이터를 선택하여 데이터프레임을 생성하고 df5 이름으로 저장 및 출력하세요.(단, '<' 와 '==' 를 반드시 사용해야 함)

In [418]:
df4.drop(columns='name', inplace=True)
df4

id_a stock_name       value   price    id_b        eps     bps     per  \
0  128940       한미약품  59385.6667  421000  130960  6301.3333   54068 15.6951   
1  130960     CJ E&M  58540.6667   98900  136480   274.1667    3551 11.4894   
2  138250      엔에스쇼핑  14558.6667   13200  138040  2122.3333   14894  6.3138   
3  139480        이마트 239230.8333  254500  139480 18268.1667  295780 13.9313   
4  142280     녹십자엠에스    468.8333   10200  145990  5741.0000  108090 14.2832   
5  145990        삼양사  82750.0000   82000  161390  5648.5000   51341  7.4533   
6  185750        종근당  40293.6667  100500  181710  2110.1667   78434 30.7559   
7  192400      쿠쿠홀딩스 179204.6667  177500  185750  3990.3333   40684 25.1859   
8  199800         툴젠  -2514.3333  115400  204210    85.1667    5335 40.8023   
9  204210     모두투어리츠   3093.3333    3475  207940  4644.1667   60099 89.7901   

     pbr  
0 1.8292  
1 0.8871  
2 0.8997  
3 0.8604  
4 0.7586  
5 0.8200  
6 0.8274  
7 2.4703  
8 0.6514  
9 6.9386

In [419]:
df5 = df4[(df4.eps < 3000) | (df4.stock_name == '이마트')]
df5

id_a stock_name       value   price    id_b        eps     bps     per  \
1  130960     CJ E&M  58540.6667   98900  136480   274.1667    3551 11.4894   
2  138250      엔에스쇼핑  14558.6667   13200  138040  2122.3333   14894  6.3138   
3  139480        이마트 239230.8333  254500  139480 18268.1667  295780 13.9313   
6  185750        종근당  40293.6667  100500  181710  2110.1667   78434 30.7559   
8  199800         툴젠  -2514.3333  115400  204210    85.1667    5335 40.8023   

     pbr  
1 0.8871  
2 0.8997  
3 0.8604  
6 0.8274  
8 0.6514

### Q20. 아래 df에서 중복 행을 제거한 후 df2에 저장하고 출력하세요.


In [421]:
import pandas as pd
df = pd.DataFrame({'c1':['a', 'a', 'b', 'a', 'b'],
                  'c2':[1, 1, 1, 2, 2],
                  'c3':[1, 1, 2, 2, 2]})
print(df)


  c1  c2  c3
0  a   1   1
1  a   1   1
2  b   1   2
3  a   2   2
4  b   2   2


In [422]:
df2 = df.drop_duplicates()
df2

c1  c2  c3
0  a   1   1
2  b   1   2
3  a   2   2
4  b   2   2

In [423]:
col1 = np.random.uniform(10, 20, size=10)
col2 =np.random.uniform(10, 20, size=10)
sr1 = pd.Series(col1, name = 'col1')
sr2 = pd.Series(col2, name = 'col2')
df = pd.concat([sr1, sr2], axis=1)
df = df.round({'col1': 2, 'col2':1})
df.head()

col1    col2
0 18.6600 18.5000
1 11.0300 19.8000
2 17.2900 17.4000
3 15.1800 16.6000
4 15.1500 14.0000

## column type 변환

In [424]:
# col2를 정수로 변환하여 출력
df1 = df.copy()
df1 = df1.col2.astype(int)
df1

0    18
1    19
2    17
3    16
4    14
5    12
6    10
7    12
8    15
9    19
Name: col2, dtype: int32

In [425]:
df2 = df.copy()
df2.iloc[1:3, :] = np.nan
df2.head()

col1    col2
0 18.6600 18.5000
1     NaN     NaN
2     NaN     NaN
3 15.1800 16.6000
4 15.1500 14.0000

In [426]:
df2.fillna(0, inplace=True)
df2

col1    col2
0 18.6600 18.5000
1  0.0000  0.0000
2  0.0000  0.0000
3 15.1800 16.6000
4 15.1500 14.0000
5 19.8400 12.4000
6 13.7300 10.6000
7 19.0800 12.6000
8 15.6700 15.8000
9 16.2400 19.9000

## 복수의 columns type 변환

In [428]:
df2 = df2.astype({'col1': 'int', 'col2':'int'})
df2.head()

col1  col2
0    18    18
1     0     0
2     0     0
3    15    16
4    15    14

In [429]:
df2['col0'] = list('abcdefghij')
df2.set_index('col0', inplace=True)
df2.head()

col1  col2
col0            
a       18    18
b        0     0
c        0     0
d       15    16
e       15    14

In [430]:
df2.isna().sum()

col1    0
col2    0
dtype: int64

In [431]:
df2 = df2.reset_index()
df2.head()

col0  col1  col2
0    a    18    18
1    b     0     0
2    c     0     0
3    d    15    16
4    e    15    14

## reindex : slicing의 개념일까?

In [432]:
df3 = df2.reindex(index = [1, 2], columns=['col1', 'col2'])
df3

col1  col2
1     0     0
2     0     0

In [433]:
df4 = df2.reindex(index = [3,4], columns=['col0', 'col2'])
df4

col0  col2
3    d    16
4    e    14

# Scaling : MinMaxScalar, StandardScalar

## MinMax

In [434]:
from sklearn.datasets import load_boston

In [444]:
boston = load_boston()
print(boston.keys())
print(boston.feature_names)
df_boston = pd.DataFrame(boston.data, 
                         columns=boston.feature_names)
df_boston['target'] = boston.target
df_boston.head()

dict_keys(['data', 'target', 'feature_names', 'DESCR', 'filename', 'data_module'])
['CRIM' 'ZN' 'INDUS' 'CHAS' 'NOX' 'RM' 'AGE' 'DIS' 'RAD' 'TAX' 'PTRATIO'
 'B' 'LSTAT']


CRIM      ZN  INDUS   CHAS    NOX     RM     AGE    DIS    RAD      TAX  \
0 0.0063 18.0000 2.3100 0.0000 0.5380 6.5750 65.2000 4.0900 1.0000 296.0000   
1 0.0273  0.0000 7.0700 0.0000 0.4690 6.4210 78.9000 4.9671 2.0000 242.0000   
2 0.0273  0.0000 7.0700 0.0000 0.4690 7.1850 61.1000 4.9671 2.0000 242.0000   
3 0.0324  0.0000 2.1800 0.0000 0.4580 6.9980 45.8000 6.0622 3.0000 222.0000   
4 0.0691  0.0000 2.1800 0.0000 0.4580 7.1470 54.2000 6.0622 3.0000 222.0000   

   PTRATIO        B  LSTAT  target  
0  15.3000 396.9000 4.9800 24.0000  
1  17.8000 396.9000 9.1400 21.6000  
2  17.8000 392.8300 4.0300 34.7000  
3  18.7000 394.6300 2.9400 33.4000  
4  18.7000 396.9000 5.3300 36.2000

In [445]:
from sklearn.preprocessing import MinMaxScaler

scalor = MinMaxScaler()
df_boston['RM'] = scalor.fit_transform(df_boston[['RM']])
df_boston[['RM']].describe()

RM
count 506.0000
mean    0.5219
std     0.1346
min     0.0000
25%     0.4454
50%     0.5073
75%     0.5868
max     1.0000

## Standard

In [446]:
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
df_boston['INDUS'] = scaler.fit_transform(df_boston[['INDUS']])
df_boston[['INDUS']].describe()

INDUS
count 506.0000
mean    0.0000
std     1.0010
min    -1.5578
25%    -0.8677
50%    -0.2111
75%     1.0160
max     2.4226

In [447]:
pd.options.display.float_format = '{:.4f}'.format

In [448]:
df_boston[['INDUS']].describe()

INDUS
count 506.0000
mean    0.0000
std     1.0010
min    -1.5578
25%    -0.8677
50%    -0.2111
75%     1.0160
max     2.4226

# encoding: LabelEncoder, OneHotEncoder

In [455]:
import pandas as pd
import numpy as np
import seaborn as sns

titanic = sns.load_dataset('titanic')

titanic['sex'] = titanic.sex.replace({'male': 0, 'female': 1} )

titanic.rename(columns = {'sex':'gender', 'age': 'age_cat', 'class': 'c_level'}, inplace = True)

titanic.columns = map(lambda x:str(x).upper(), titanic.columns)

titanic.rename(columns=str.lower, inplace=True)

df = titanic[['survived', 'pclass', 'gender', 'c_level']]
df.head()

survived  pclass  gender c_level
0         0       3       0   Third
1         1       1       1   First
2         1       3       1   Third
3         1       1       1   First
4         0       3       0   Third

In [456]:
df['gender'].replace([1, 0], ['female', 'male'], inplace=True)
df.head()

survived  pclass  gender c_level
0         0       3    male   Third
1         1       1  female   First
2         1       3  female   Third
3         1       1  female   First
4         0       3    male   Third

## encoding

In [457]:
from sklearn.preprocessing import LabelEncoder

df1=df.copy()
le = LabelEncoder()
features = ['gender', 'c_level']
for f in features:
    df1[f] = le.fit_transform(df1[f])
df1.head()

survived  pclass  gender  c_level
0         0       3       1        2
1         1       1       0        0
2         1       3       0        2
3         1       1       0        0
4         0       3       1        2

알파벳순인듯
before after 알 수 없음?

In [458]:
df1.c_level.value_counts()

2    491
0    216
1    184
Name: c_level, dtype: int64

## 인코딩- OneHotEncoder

```
희소행렬(<->밀집행렬)
masking 같은 느낌인듯
```

In [459]:
df2 = df.copy()
df2.head()

survived  pclass  gender c_level
0         0       3    male   Third
1         1       1  female   First
2         1       3  female   Third
3         1       1  female   First
4         0       3    male   Third

In [462]:
df3 = df2[['gender', 'c_level']]
pd.get_dummies(df3).head() # 1, 0으로 

gender_female  gender_male  c_level_First  c_level_Second  c_level_Third
0              0            1              0               0              1
1              1            0              1               0              0
2              1            0              0               0              1
3              1            0              1               0              0
4              0            1              0               0              1

## (D) 과제 #1 Tue 0705

 - titanic 데이터셋에서 enbarked, who, adult_male 에 대하여 LableEncoding하세요.
 - titanic 데이터셋에서 enbarked, who, adult_male 에 대하여 OneHotEncoding하세요.

 - titanic 데이터셋에서 enbarked, who, adult_male 에 대하여 LableEncoding하세요.

In [463]:
titanic = sns.load_dataset('titanic')
titanic.head()

survived  pclass     sex     age  sibsp  parch    fare embarked  class  \
0         0       3    male 22.0000      1      0  7.2500        S  Third   
1         1       1  female 38.0000      1      0 71.2833        C  First   
2         1       3  female 26.0000      0      0  7.9250        S  Third   
3         1       1  female 35.0000      1      0 53.1000        S  First   
4         0       3    male 35.0000      0      0  8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True

In [464]:
df_l=titanic.copy()
le = LabelEncoder()
features = ['embarked', 'who', 'adult_male']
for f in features:
    df_l[f] = le.fit_transform(df_l[f])
df_l.head()

survived  pclass     sex     age  sibsp  parch    fare  embarked  class  \
0         0       3    male 22.0000      1      0  7.2500         2  Third   
1         1       1  female 38.0000      1      0 71.2833         0  First   
2         1       3  female 26.0000      0      0  7.9250         2  Third   
3         1       1  female 35.0000      1      0 53.1000         2  First   
4         0       3    male 35.0000      0      0  8.0500         2  Third   

   who  adult_male deck  embark_town alive  alone  
0    1           1  NaN  Southampton    no  False  
1    2           0    C    Cherbourg   yes  False  
2    2           0  NaN  Southampton   yes   True  
3    2           0    C  Southampton   yes  False  
4    1           1  NaN  Southampton    no   True

 - titanic 데이터셋에서 enbarked, who, adult_male 에 대하여 OneHotEncoding하세요.

In [465]:
df_o = titanic[['embarked', 'who', 'adult_male']]
pd.get_dummies(df_o).head()

adult_male  embarked_C  embarked_Q  embarked_S  who_child  who_man  \
0        True           0           0           1          0        1   
1       False           1           0           0          0        0   
2       False           0           0           1          0        0   
3       False           0           0           1          0        0   
4        True           0           0           1          0        1   

   who_woman  
0          0  
1          1  
2          1  
3          1  
4          0

# (D) 과제 #6 Tue 0705

## Q21. Q20의 df에서 drop_duplicates() 메소드를 사용하여 c2, c3열을 기준으로 중복 행을 제거한 후 df3에 저장하고 출력하세요.


In [466]:
df = pd.DataFrame({'c1':['a', 'a', 'b', 'a', 'b'],
                  'c2':[1, 1, 1, 2, 2],
                  'c3':[1, 1, 2, 2, 2]})
df

c1  c2  c3
0  a   1   1
1  a   1   1
2  b   1   2
3  a   2   2
4  b   2   2

In [467]:
df3 = df.drop_duplicates(subset=['c2', 'c3'])
df3

c1  c2  c3
0  a   1   1
2  b   1   2
3  a   2   2

## Q33. './dataset/stock-data.csv'를 데이터프레임으로 불러와서 datetime64 자료형으로 변환한 후에 년, 월, 일로 분리하고 year를 인덱스로 셋팅하여 출력하세요.


In [470]:
# 데이터프레임으로 불러와서
df_stock = pd.read_csv('./dataset/stock-data.csv')
df_stock.head()

Date  Close  Start   High    Low  Volume
0  2018-07-02  10100  10850  10900  10000  137977
1  2018-06-29  10700  10550  10900   9990  170253
2  2018-06-28  10400  10900  10950  10150  155769
3  2018-06-27  10900  10800  11050  10500  133548
4  2018-06-26  10800  10900  11000  10700   63039

In [471]:
# datetime64로 자료형을 변환
from datetime import datetime

In [472]:
def to_datetime(x):
    date = datetime.strptime(x, '%Y-%m-%d')
    return date

In [473]:
df_stock1 = df_stock.copy()
df_stock1.Date = df_stock1.Date.apply(to_datetime)

In [475]:
df_stock1.Date.info()

<class 'pandas.core.series.Series'>
RangeIndex: 20 entries, 0 to 19
Series name: Date
Non-Null Count  Dtype         
--------------  -----         
20 non-null     datetime64[ns]
dtypes: datetime64[ns](1)
memory usage: 288.0 bytes


In [476]:
# 년, 월, 일 분리

In [477]:
df_stock1['year'] = df_stock1.Date.apply(lambda x: x.year)
df_stock1['month'] = df_stock1.Date.apply(lambda x: x.month)
df_stock1['day'] = df_stock1.Date.apply(lambda x: x.day)

df_stock1.head()

Date  Close  Start   High    Low  Volume  year  month  day
0 2018-07-02  10100  10850  10900  10000  137977  2018      7    2
1 2018-06-29  10700  10550  10900   9990  170253  2018      6   29
2 2018-06-28  10400  10900  10950  10150  155769  2018      6   28
3 2018-06-27  10900  10800  11050  10500  133548  2018      6   27
4 2018-06-26  10800  10900  11000  10700   63039  2018      6   26

In [478]:
# year을 인덱스로 셋팅하고 출력
df_stock1.set_index('year', inplace=True)
df_stock1.head()

Date  Close  Start   High    Low  Volume  month  day
year                                                           
2018 2018-07-02  10100  10850  10900  10000  137977      7    2
2018 2018-06-29  10700  10550  10900   9990  170253      6   29
2018 2018-06-28  10400  10900  10950  10150  155769      6   28
2018 2018-06-27  10900  10800  11050  10500  133548      6   27
2018 2018-06-26  10800  10900  11000  10700   63039      6   26

## Q34. './dataset/stock-data.csv'를 데이터프레임으로 불러와서 datetime64 자료형으로 변환한 후에 Period로 변환하고 년월일 표기 변경하여 년월을 인데스로 셋팅하여 출력하세요.


In [479]:
# 데이터프레임으로 불러와서
df_stock = pd.read_csv('./dataset/stock-data.csv')
df_stock.head()

Date  Close  Start   High    Low  Volume
0  2018-07-02  10100  10850  10900  10000  137977
1  2018-06-29  10700  10550  10900   9990  170253
2  2018-06-28  10400  10900  10950  10150  155769
3  2018-06-27  10900  10800  11050  10500  133548
4  2018-06-26  10800  10900  11000  10700   63039

In [480]:
# datetime64 자료형으로 변환

In [481]:
def to_datetime(x):
    date = datetime.strptime(x, '%Y-%m-%d')
    return date

In [482]:
df_stock1 = df_stock.copy()
df_stock1.Date = df_stock1.Date.apply(to_datetime)

In [483]:
df_stock1.Date.info()

<class 'pandas.core.series.Series'>
RangeIndex: 20 entries, 0 to 19
Series name: Date
Non-Null Count  Dtype         
--------------  -----         
20 non-null     datetime64[ns]
dtypes: datetime64[ns](1)
memory usage: 288.0 bytes


In [484]:
# Period로 변환
def to_period(x):
    dt_pd = pd.to_datetime(x)
    pr = dt_pd.to_period(freq='M')
    return pr

In [485]:
df_stock2 = df_stock1.copy()
df_stock2.Date = df_stock2.Date.apply(to_period)
df_stock2.info()
df_stock2.Date.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20 entries, 0 to 19
Data columns (total 6 columns):
 #   Column  Non-Null Count  Dtype    
---  ------  --------------  -----    
 0   Date    20 non-null     period[M]
 1   Close   20 non-null     int64    
 2   Start   20 non-null     int64    
 3   High    20 non-null     int64    
 4   Low     20 non-null     int64    
 5   Volume  20 non-null     int64    
dtypes: int64(5), period[M](1)
memory usage: 1.1 KB


0    2018-07
1    2018-06
2    2018-06
3    2018-06
4    2018-06
Name: Date, dtype: period[M]

- period
    - dt_pd = pd.to_datetime(datetime) #여기서 arg로 들어간 datetime은 datetime 것
    - pr = datetime.to_period(feq='M')

# ★ Period Test

In [487]:
date = datetime(2022, 3, 15)
date

datetime.datetime(2022, 3, 15, 0, 0)

In [488]:
pd_dt = pd.to_datetime(date)
pd_dt, type(pd_dt)

(Timestamp('2022-03-15 00:00:00'), pandas._libs.tslibs.timestamps.Timestamp)

In [489]:
pr = pd_dt.to_period(freq='M')
pr, type(pr)

(Period('2022-03', 'M'), pandas._libs.tslibs.period.Period)

In [490]:
pr.year, type(pr.year)

(2022, int)